// @larksuite/channel/dist/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
//#region \0rolldown/runtime.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
		key = keys[i];
		if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
			get: ((k) => from[k]).bind(null, key),
			enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
		});
	}
	return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
	value: mod,
	enumerable: true
}) : target, mod));
//#endregion
let node_fs = require("node:fs");
let node_fs_promises = require("node:fs/promises");
let node_stream_promises = require("node:stream/promises");
let _larksuiteoapi_node_sdk = require("@larksuiteoapi/node-sdk");
let https_proxy_agent = require("https-proxy-agent");
let fs = require("fs");
fs = __toESM(fs, 1);
let http = require("http");
http = __toESM(http, 1);
let https = require("https");
https = __toESM(https, 1);
let path = require("path");
path = __toESM(path, 1);
let dns = require("dns");
let net = require("net");
//#region src/chat-mode-cache.ts
/**
* In-memory cache for chat mode lookups. Feishu omits chat mode from message
* events, so the only way to know whether a chat is a p2p / ordinary group /
* topic group is an extra `chat.get` — which is stable for a chat's lifetime,
* hence cacheable by chatId.
*
* On lookup failure (network / permission / unknown chatId) the resolver
* returns 'group' (the conservative default: treat as an ordinary chat) and
* does NOT poison the cache, so a later message gets another try.
*/
var ChatModeCache = class {
	cache = /* @__PURE__ */ new Map();
	async resolve(chatId, fetch) {
		const hit = this.cache.get(chatId);
		if (hit) return hit;
		try {
			const mode = await fetch(chatId);
			this.cache.set(chatId, mode);
			return mode;
		} catch {
			return "group";
		}
	}
	invalidate(chatId) {
		this.cache.delete(chatId);
	}
};
//#endregion
//#region src/comments.ts
/** File types the drive comment APIs support here. Others (slides, bitable,
*  mindnote) use different APIs and are out of scope. */
const SUPPORTED_FILE_TYPES = new Set([
	"doc",
	"docx",
	"sheet",
	"file"
]);
function errCode(err) {
	return err?.response?.data?.code;
}
var CommentSurface = class {
	client;
	logger;
	constructor(client, logger) {
		this.client = client;
		this.logger = logger;
	}
	/**
	* Resolve the (fileToken, fileType) to hit for the comment APIs. If the
	* token is a wiki node, swap to its underlying obj_token; otherwise pass
	* through. Returns `null` when the file type is unsupported.
	*/
	async resolveTarget(fileToken, fileType) {
		if (!SUPPORTED_FILE_TYPES.has(fileType)) return null;
		const passthrough = {
			fileToken,
			fileType
		};
		try {
			const node = (await this.client.wiki.v2.space.getNode({ params: { token: fileToken } }))?.data?.node;
			if (node?.obj_token && node.obj_type && SUPPORTED_FILE_TYPES.has(node.obj_type)) {
				this.logger.info?.("channel: comment wiki-resolved", {
					objToken: node.obj_token,
					objType: node.obj_type
				});
				return {
					fileToken: node.obj_token,
					fileType: node.obj_type
				};
			}
		} catch {}
		return passthrough;
	}
	/**
	* Fetch a comment with its replies. Tries `fileComment.get`; for comment
	* types that return 1069307 there, falls back to paginating `.list` and
	* locating the comment by id. Returns `null` when not found.
	*/
	async fetch(target, commentId) {
		try {
			const r = await this.client.drive.v1.fileComment.get({
				params: { file_type: target.fileType },
				path: {
					file_token: target.fileToken,
					comment_id: commentId
				}
			});
			return {
				commentId,
				replies: r?.data?.reply_list?.replies ?? [],
				quote: r?.data?.quote || void 0,
				isWhole: Boolean(r?.data?.is_whole)
			};
		} catch (err) {
			this.logger.warn?.("channel: comment get failed, falling back to list", { code: errCode(err) });
			const found = await this.findViaList(target, commentId);
			if (!found) return null;
			return {
				commentId,
				replies: found.reply_list?.replies ?? [],
				quote: found.quote || void 0,
				isWhole: Boolean(found.is_whole)
			};
		}
	}
	/**
	* Reply to a comment in-thread. Whole-document comments reject in-thread
	* replies with 1069302 (they have no thread, only a flat list) — in that
	* case post a fresh top-level comment instead.
	*
	* When the caller already knows the comment is whole-document (e.g. from
	* {@link FetchedComment.isWhole}), pass `{ topLevel: true }` to skip the
	* doomed in-thread probe and post top-level directly — saving one
	* round-trip. See {@link replyTopLevel} for the explicit form.
	*/
	async reply(target, commentId, text, opts) {
		if (opts?.topLevel) {
			await this.replyTopLevel(target, text);
			return;
		}
		const url = `/open-apis/drive/v1/files/${encodeURIComponent(target.fileToken)}/comments/${encodeURIComponent(commentId)}/replies?file_type=${encodeURIComponent(target.fileType)}`;
		try {
			await this.client.request({
				method: "POST",
				url,
				data: { content: { elements: [{
					type: "text_run",
					text_run: { text }
				}] } }
			});
			this.logger.info?.("channel: comment replied", { mode: "in-thread" });
			return;
		} catch (err) {
			if (errCode(err) !== 1069302) throw err;
			this.logger.warn?.("channel: comment reply rejected, posting fresh top-level", { code: 1069302 });
		}
		await this.replyTopLevel(target, text);
		this.logger.info?.("channel: comment replied", { mode: "new-top-level" });
	}
	/**
	* Post a fresh top-level comment on the file (no in-thread probe). Use this
	* when you already know in-thread replies won't apply — chiefly
	* whole-document comments, where {@link reply} would otherwise waste a
	* round-trip discovering 1069302.
	*/
	async replyTopLevel(target, text) {
		await this.client.drive.v1.fileComment.create({
			params: { file_type: target.fileType },
			path: { file_token: target.fileToken },
			data: { reply_list: { replies: [{ content: { elements: [{
				type: "text_run",
				text_run: { text }
			}] } }] } }
		});
	}
	/**
	* Add a reaction to a comment reply. Doc-comment reactions use a dedicated
	* endpoint (separate from IM message reactions); add/delete are the same
	* POST distinguished by an `action` field, and it returns no reaction_id.
	* Returns `true` on success. Defaults to the "Typing" emoji.
	*/
	async addReaction(target, replyId, emojiType = "Typing") {
		return this.reaction(target, replyId, emojiType, "add");
	}
	/** Remove a previously-added comment reaction. Same endpoint, action=delete. */
	async removeReaction(target, replyId, emojiType = "Typing") {
		await this.reaction(target, replyId, emojiType, "delete");
	}
	async reaction(target, replyId, emojiType, action) {
		const url = `/open-apis/drive/v2/files/${encodeURIComponent(target.fileToken)}/comments/reaction?file_type=${encodeURIComponent(target.fileType)}`;
		try {
			await this.client.request({
				method: "POST",
				url,
				data: {
					action,
					reply_id: replyId,
					reaction_type: emojiType
				}
			});
			this.logger.info?.(`channel: comment reaction ${action}`, {
				fileToken: target.fileToken,
				replyId
			});
			return true;
		} catch (err) {
			this.logger.warn?.(`channel: comment reaction ${action} failed`, {
				fileToken: target.fileToken,
				replyId,
				err: err instanceof Error ? err.message : String(err)
			});
			return false;
		}
	}
	async findViaList(target, commentId) {
		let pageToken;
		for (let page = 0; page < 10; page++) {
			const r = await this.client.drive.v1.fileComment.list({
				params: {
					file_type: target.fileType,
					page_size: 100,
					...pageToken ? { page_token: pageToken } : {}
				},
				path: { file_token: target.fileToken }
			});
			const hit = (r?.data?.items ?? []).find((it) => it.comment_id === commentId);
			if (hit) return hit;
			if (!r?.data?.has_more || !r.data.page_token) break;
			pageToken = r.data.page_token;
		}
		return null;
	}
};
//#endregion
//#region src/keepalive.ts
/**
* App-level keepalive loop — defense-in-depth against silent WS / network
* issues the SDK's internal ping watchdog might miss. Sunk from bridge's
* `bot/keepalive.ts`.
*
*  1. Independent timer (default 15s), untied from the server-pushed ping
*     cadence, to catch issues earlier and from a different angle.
*  2. Wake-up detection — if the timer was skipped for > SLEEP_DETECT_MS the
*     machine likely slept; reset counters and bail for this tick.
*  3. Timer-storm guard — on wake, multiple intervals can fire back-to-back.
*  4. HTTP probe — before force-reconnecting, check the Feishu domain is
*     reachable; if not it's a network outage, not a WS problem.
*  5. Counter-based debounce — only force-reconnect after DEAD_THRESHOLD
*     consecutive ticks confirm WS is not connected.
*
* The "what to do when reconnect can't recover" policy stays with the app via
* `onUnrecoverable` (e.g. restart the process).
*/
const DEFAULT_INTERVAL_MS = 15e3;
const SLEEP_DETECT_MS = 3e4;
const TIMER_STORM_GUARD_MS = 5e3;
const HTTP_PROBE_TIMEOUT_MS = 5e3;
const DEAD_THRESHOLD = 3;
const NETWORK_DOWN_LOG_EVERY = 20;
function startKeepalive(deps) {
	const { getConnectionStatus, domain, forceReconnect, onUnrecoverable, logger } = deps;
	const intervalMs = deps.intervalMs ?? DEFAULT_INTERVAL_MS;
	let lastTick = 0;
	let consecutiveDown = 0;
	let networkDownTicks = 0;
	let stopped = false;
	const tick = async () => {
		if (stopped) return;
		const now = Date.now();
		const sinceLast = lastTick > 0 ? now - lastTick : 0;
		if (sinceLast > 0 && sinceLast < TIMER_STORM_GUARD_MS) return;
		if (sinceLast > SLEEP_DETECT_MS) {
			logger.info?.("channel: keepalive wake-up", { sleptMs: sinceLast });
			consecutiveDown = 0;
			networkDownTicks = 0;
			lastTick = now;
			return;
		}
		lastTick = now;
		const status = getConnectionStatus();
		if (!status) return;
		if (status.state === "connected") {
			if (consecutiveDown > 0) logger.info?.("channel: keepalive recovered", { afterTicks: consecutiveDown });
			consecutiveDown = 0;
			networkDownTicks = 0;
			return;
		}
		if (!await httpProbe(domain)) {
			networkDownTicks++;
			if (networkDownTicks === 1 || networkDownTicks % NETWORK_DOWN_LOG_EVERY === 0) logger.warn?.("channel: network unreachable", {
				domain,
				networkDownTicks
			});
			consecutiveDown = 0;
			return;
		}
		if (networkDownTicks > 0) {
			logger.info?.("channel: network reachable again", { afterTicks: networkDownTicks });
			networkDownTicks = 0;
		}
		consecutiveDown++;
		logger.warn?.("channel: keepalive ws-stuck", {
			state: status.state,
			reconnectAttempts: status.reconnectAttempts,
			consecutiveDown
		});
		if (consecutiveDown >= DEAD_THRESHOLD) {
			logger.warn?.("channel: keepalive force-reconnect", { state: status.state });
			consecutiveDown = 0;
			try {
				await forceReconnect();
			} catch (err) {
				logger.error?.("channel: keepalive force-reconnect failed", err);
				onUnrecoverable?.(err);
			}
		}
	};
	const timer = setInterval(() => {
		tick().catch((err) => logger.error?.("channel: keepalive tick failed", err));
	}, intervalMs);
	timer.unref?.();
	return { stop() {
		stopped = true;
		clearInterval(timer);
	} };
}
async function httpProbe(domain) {
	const target = /^https?:\/\//i.test(domain) ? domain : "https://open.feishu.cn";
	try {
		const ctrl = new AbortController();
		const timer = setTimeout(() => ctrl.abort(), HTTP_PROBE_TIMEOUT_MS);
		try {
			return (await fetch(target, {
				method: "HEAD",
				signal: ctrl.signal
			})).status > 0;
		} finally {
			clearTimeout(timer);
		}
	} catch {
		return false;
	}
}
//#endregion
//#region src/normalize/mentions.ts
function isMentionAll(m) {
	return m.key === "@_all";
}
function extractMentions(raw, botOpenId) {
	const mentions = /* @__PURE__ */ new Map();
	const mentionsByOpenId = /* @__PURE__ */ new Map();
	const mentionList = [];
	let mentionAll = false;
	let mentionedBot = false;
	for (const m of raw ?? []) {
		if (isMentionAll(m)) {
			mentionAll = true;
			mentions.set(m.key, {
				key: m.key,
				name: m.name,
				isBot: false
			});
			continue;
		}
		const openId = m.id?.open_id ?? "";
		const userId = m.id?.user_id;
		const isBot = Boolean(botOpenId && openId === botOpenId);
		if (isBot) mentionedBot = true;
		const info = {
			key: m.key,
			openId: openId || void 0,
			userId,
			name: m.name,
			isBot
		};
		mentions.set(m.key, info);
		if (openId) mentionsByOpenId.set(openId, info);
		mentionList.push(info);
	}
	return {
		mentions,
		mentionsByOpenId,
		mentionList,
		mentionAll,
		mentionedBot
	};
}
/**
* Second-pass: replace placeholder keys in `content` with human-readable names
* (or strip bot mentions if configured).
*
* Must run AFTER all converters have done their work, because converters use
* placeholder keys to defer resolution to this single point.
*/
function resolveMentions(content, ctx) {
	if (!content || ctx.mentions.size === 0) return content;
	let out = content;
	for (const [key, info] of ctx.mentions) {
		if (info.isBot && ctx.stripBotMentions) {
			const re = new RegExp(`\\s?${escapeRegex(key)}\\s?`, "g");
			out = out.replace(re, " ");
			continue;
		}
		const replacement = info.name ? `@${info.name}` : key;
		out = out.split(key).join(replacement);
	}
	return out.replace(/[ \t]{2,}/g, " ").trim();
}
function escapeRegex(s) {
	return s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}
//#endregion
//#region src/normalize/utils.ts
function safeParse(raw) {
	if (!raw) return void 0;
	try {
		return JSON.parse(raw);
	} catch {
		return;
	}
}
function applyStyle(text, style) {
	if (!style || style.length === 0) return text;
	let out = text;
	if (style.includes("bold")) out = `**${out}**`;
	if (style.includes("italic")) out = `*${out}*`;
	if (style.includes("underline")) out = `<u>${out}</u>`;
	if (style.includes("lineThrough") || style.includes("strikethrough")) out = `~~${out}~~`;
	if (style.includes("codeInline") || style.includes("code")) out = `\`${out}\``;
	return out;
}
const LOCALE_PRIORITY = [
	"zh_cn",
	"en_us",
	"ja_jp"
];
function unwrapLocale(parsed) {
	if ("title" in parsed || "content" in parsed) return parsed;
	for (const loc of LOCALE_PRIORITY) {
		const hit = parsed[loc];
		if (hit != null && typeof hit === "object") return hit;
	}
	const firstKey = Object.keys(parsed)[0];
	if (firstKey) {
		const first = parsed[firstKey];
		if (first != null && typeof first === "object") return first;
	}
}
function formatDuration(ms) {
	if (ms == null || !Number.isFinite(ms) || ms < 0) return void 0;
	if (ms < 1e3) return `${Math.round(ms)}ms`;
	if (ms % 1e3 === 0) return `${ms / 1e3}s`;
	return `${(ms / 1e3).toFixed(1)}s`;
}
function millisToDatetime(ms) {
	if (ms == null) return void 0;
	const n = typeof ms === "string" ? parseInt(ms, 10) : ms;
	if (!Number.isFinite(n) || n <= 0) return void 0;
	const d = new Date(n + 8 * 36e5);
	return `${d.getUTCFullYear()}-${String(d.getUTCMonth() + 1).padStart(2, "0")}-${String(d.getUTCDate()).padStart(2, "0")} ${String(d.getUTCHours()).padStart(2, "0")}:${String(d.getUTCMinutes()).padStart(2, "0")}`;
}
function formatRFC3339Beijing(ms) {
	const d = new Date(ms + 8 * 36e5);
	return `${d.getUTCFullYear()}-${String(d.getUTCMonth() + 1).padStart(2, "0")}-${String(d.getUTCDate()).padStart(2, "0")}T${String(d.getUTCHours()).padStart(2, "0")}:${String(d.getUTCMinutes()).padStart(2, "0")}:${String(d.getUTCSeconds()).padStart(2, "0")}+08:00`;
}
function indentLines(text, indent) {
	return text.split("\n").map((line) => `${indent}${line}`).join("\n");
}
function escapeAttr(s) {
	return s.replace(/"/g, "&quot;");
}
//#endregion
//#region src/normalize/converters/audio.ts
const convertAudio = async (raw, _ctx) => {
	const parsed = safeParse(raw);
	const fileKey = parsed?.file_key;
	if (!fileKey) return {
		content: "[audio]",
		resources: []
	};
	const duration = parsed?.duration;
	const durAttr = formatDuration(duration);
	return {
		content: `<audio key="${fileKey}"${durAttr ? ` duration="${durAttr}"` : ""}/>`,
		resources: [{
			type: "audio",
			fileKey,
			durationMs: duration
		}]
	};
};
//#endregion
//#region src/normalize/converters/calendar.ts
function formatCalendarInner(raw) {
	const parsed = safeParse(raw);
	if (!parsed) return "[calendar event]";
	const lines = [];
	if (parsed.summary) lines.push(`📅 ${parsed.summary}`);
	const start = millisToDatetime(parsed.start_time);
	const end = millisToDatetime(parsed.end_time);
	if (start && end) lines.push(`🕙 ${start} ~ ${end}`);
	else if (start) lines.push(`🕙 ${start}`);
	return lines.length > 0 ? lines.join("\n") : "[calendar event]";
}
const convertCalendar = async (raw, _ctx) => ({
	content: `<calendar_invite>\n${formatCalendarInner(raw)}\n</calendar_invite>`,
	resources: []
});
const convertGeneralCalendar = async (raw, _ctx) => ({
	content: `<calendar>\n${formatCalendarInner(raw)}\n</calendar>`,
	resources: []
});
const convertShareCalendarEvent = async (raw, _ctx) => ({
	content: `<calendar_share>\n${formatCalendarInner(raw)}\n</calendar_share>`,
	resources: []
});
//#endregion
//#region src/normalize/converters/fallback.ts
const convertUnknown = async (raw, _ctx) => {
	const parsed = safeParse(raw);
	if (parsed && typeof parsed.text === "string") return {
		content: parsed.text,
		resources: []
	};
	return {
		content: "[unsupported message]",
		resources: []
	};
};
//#endregion
//#region src/normalize/converters/file.ts
const convertFile = async (raw, _ctx) => {
	const parsed = safeParse(raw);
	const fileKey = parsed?.file_key;
	if (!fileKey) return {
		content: "[file]",
		resources: []
	};
	const fileName = parsed?.file_name;
	return {
		content: `<file key="${fileKey}"${fileName ? ` name="${escapeAttr(fileName)}"` : ""}/>`,
		resources: [{
			type: "file",
			fileKey,
			fileName
		}]
	};
};
//#endregion
//#region src/normalize/converters/folder.ts
const convertFolder = async (raw, _ctx) => {
	const parsed = safeParse(raw);
	const fileKey = parsed?.file_key;
	if (!fileKey) return {
		content: "[folder]",
		resources: []
	};
	return {
		content: `<folder key="${fileKey}"${parsed?.file_name ? ` name="${escapeAttr(parsed.file_name)}"` : ""}/>`,
		resources: []
	};
};
//#endregion
//#region src/normalize/converters/hongbao.ts
const convertHongbao = async (raw, _ctx) => {
	const parsed = safeParse(raw);
	return {
		content: `<hongbao${parsed?.text ? ` text="${escapeAttr(parsed.text)}"` : ""}/>`,
		resources: []
	};
};
//#endregion
//#region src/normalize/converters/image.ts
const convertImage = async (raw, _ctx) => {
	const imageKey = safeParse(raw)?.image_key;
	if (!imageKey) return {
		content: "[image]",
		resources: []
	};
	const resources = [{
		type: "image",
		fileKey: imageKey
	}];
	return {
		content: `![image](${imageKey})`,
		resources
	};
};
//#endregion
//#region src/normalize/converters/interactive/card-walker.ts
/**
* Recursive walker that extracts human-readable text from a Feishu
* interactive card JSON tree. Covers header titles, plain_text / lark_md
* elements, button labels, form fields, notes, and commonly used nested
* element types.
*/
function walkCard(node) {
	const out = [];
	visit(node, out);
	return out.filter((s) => s && s.trim().length > 0);
}
function visit(node, out) {
	if (node == null) return;
	if (typeof node === "string" || typeof node === "number" || typeof node === "boolean") return;
	if (Array.isArray(node)) {
		for (const child of node) visit(child, out);
		return;
	}
	if (typeof node !== "object") return;
	const obj = node;
	const tag = obj.tag;
	if (typeof tag === "string" && (tag === "plain_text" || tag === "lark_md" || tag === "markdown")) {
		if (typeof obj.content === "string") out.push(obj.content);
		return;
	}
	if (obj.header && typeof obj.header === "object") {
		const header = obj.header;
		if (header.title) visit(header.title, out);
	}
	if (obj.text) visit(obj.text, out);
	if (typeof tag === "string" && tag === "button") {
		const text = obj.text;
		if (text) visit(text, out);
	}
	if (obj.label) visit(obj.label, out);
	if (obj.placeholder) visit(obj.placeholder, out);
	if (Array.isArray(obj.options)) for (const opt of obj.options) {
		const o = opt;
		if (o?.text) visit(o.text, out);
	}
	if (Array.isArray(obj.elements)) for (const el of obj.elements) visit(el, out);
	if (Array.isArray(obj.fields)) for (const f of obj.fields) visit(f, out);
	if (Array.isArray(obj.actions)) for (const a of obj.actions) visit(a, out);
	if (Array.isArray(obj.columns)) for (const c of obj.columns) visit(c, out);
	if (obj.body) visit(obj.body, out);
}
//#endregion
//#region src/normalize/converters/interactive/index.ts
const convertInteractive = async (raw, _ctx) => {
	const parsed = safeParse(raw);
	if (parsed == null || typeof parsed !== "object") return {
		content: "[interactive card]",
		resources: []
	};
	const pieces = walkCard(parsed);
	if (pieces.length === 0) return {
		content: "[interactive card]",
		resources: []
	};
	const seen = /* @__PURE__ */ new Set();
	const out = [];
	for (const p of pieces) {
		const key = p.trim();
		if (!key || seen.has(key)) continue;
		seen.add(key);
		out.push(key);
	}
	return {
		content: out.join("\n"),
		resources: []
	};
};
//#endregion
//#region src/normalize/converters/location.ts
const convertLocation = async (raw, _ctx) => {
	const parsed = safeParse(raw);
	const name = parsed?.name;
	const lat = parsed?.latitude;
	const lng = parsed?.longitude;
	return {
		content: `<location${name ? ` name="${escapeAttr(name)}"` : ""}${lat && lng ? ` coords="lat:${lat},lng:${lng}"` : ""}/>`,
		resources: []
	};
};
//#endregion
//#region src/normalize/converters/merge-forward.ts
const MAX_ITEMS = 50;
const convertMergeForward = async (_raw, ctx) => {
	const { messageId, fetchSubMessages, dispatch } = ctx;
	if (!fetchSubMessages || !dispatch) return {
		content: "<forwarded_messages/>",
		resources: []
	};
	let items;
	try {
		items = await fetchSubMessages(messageId);
	} catch {
		return {
			content: "<forwarded_messages/>",
			resources: []
		};
	}
	if (!items || items.length === 0) return {
		content: "<forwarded_messages/>",
		resources: []
	};
	const capped = items.slice(0, MAX_ITEMS);
	const truncated = items.length > MAX_ITEMS;
	if (ctx.batchResolveNames) {
		const senderIds = /* @__PURE__ */ new Set();
		for (const it of capped) {
			const sid = it.sender?.id;
			if (sid && it.message_id !== messageId) senderIds.add(sid);
		}
		if (senderIds.size > 0) try {
			await ctx.batchResolveNames([...senderIds]);
		} catch {}
	}
	const { content, resources } = await formatSubTree(messageId, buildChildrenMap(capped, messageId), ctx, truncated);
	return {
		content,
		resources
	};
};
function buildChildrenMap(items, rootId) {
	const map = /* @__PURE__ */ new Map();
	for (const it of items) {
		if (it.message_id === rootId && !it.upper_message_id) continue;
		const pid = it.upper_message_id ?? rootId;
		let arr = map.get(pid);
		if (!arr) {
			arr = [];
			map.set(pid, arr);
		}
		arr.push(it);
	}
	for (const arr of map.values()) arr.sort((a, b) => {
		return parseInt(String(a.create_time ?? "0"), 10) - parseInt(String(b.create_time ?? "0"), 10);
	});
	return map;
}
async function formatSubTree(parentId, map, ctx, truncated = false) {
	const children = map.get(parentId);
	if (!children || children.length === 0) return {
		content: "<forwarded_messages/>",
		resources: []
	};
	const parts = [];
	const resources = [];
	for (const item of children) try {
		const sub = await renderItem(item, map, ctx);
		if (sub.content) parts.push(sub.content);
		resources.push(...sub.resources);
	} catch {}
	if (parts.length === 0) return {
		content: "<forwarded_messages/>",
		resources
	};
	return {
		content: `<forwarded_messages>\n${parts.join("\n")}${truncated ? "\n... (truncated)" : ""}\n</forwarded_messages>`,
		resources
	};
}
async function renderItem(item, map, ctx) {
	const msgType = item.msg_type ?? "text";
	const senderId = item.sender?.id ?? "unknown";
	const createMs = parseInt(String(item.create_time ?? "0"), 10);
	const timestamp = createMs > 0 ? formatRFC3339Beijing(createMs) : "unknown";
	const displayName = ctx.resolveUserName?.(senderId) ?? senderId;
	let content;
	let resources = [];
	if (msgType === "merge_forward") {
		const nestedId = item.message_id;
		if (nestedId) {
			const sub = await formatSubTree(nestedId, map, ctx);
			content = sub.content;
			resources = sub.resources;
		} else content = "<forwarded_messages/>";
	} else {
		const rawContent = item.body?.content ?? "{}";
		if (!ctx.dispatch) content = rawContent;
		else {
			const r = await ctx.dispatch(rawContent, msgType, ctx);
			content = r.content;
			resources = r.resources;
		}
	}
	return {
		content: `[${timestamp}] ${displayName}:\n${indentLines(content, "    ")}`,
		resources
	};
}
//#endregion
//#region src/normalize/converters/post.ts
const atMentionRe = /<at(\s+)user_id(\s*)=(\s*)"(.*?)">(.*?)<\/at>/g;
const imageKeyRe = /!\[(.*?)\]\(([^)]+)\)/g;
const convertPost = async (raw, ctx) => {
	const rawParsed = safeParse(raw);
	if (rawParsed == null || typeof rawParsed !== "object") return {
		content: "[rich text message]",
		resources: []
	};
	const body = unwrapLocale(rawParsed);
	if (!body) return {
		content: "[rich text message]",
		resources: []
	};
	const sourceParagraphs = body.content_v2 && body.content_v2.length > 0 ? body.content_v2 : body.content ?? [];
	const resources = [];
	const lines = [];
	if (body.title) {
		lines.push(`**${body.title}**`);
		lines.push("");
	}
	for (const paragraph of sourceParagraphs) {
		if (!Array.isArray(paragraph)) continue;
		let line = "";
		for (const el of paragraph) line += renderElement(el, ctx, resources);
		lines.push(line);
	}
	return {
		content: lines.join("\n").trim() || "[rich text message]",
		resources
	};
};
/**
* Post-process raw markdown text from an "md" element.
* Splits by fenced code block delimiters (```) and only applies
* transformations to text outside of properly paired code blocks.
* Unclosed fences are treated as outside-code-block text.
*/
function processMdText(text, resources) {
	const parts = text.split("```");
	const total = parts.length;
	for (let i = 0; i < parts.length; i++) {
		let isInside = i % 2 === 1;
		if (isInside && total % 2 === 0 && i === total - 1) isInside = false;
		if (!isInside) {
			parts[i] = parts[i].replace(atMentionRe, (_match, _sp1, _sp2, _sp3, userId, name) => {
				if (userId === "all" || userId === "all_members") return "@all";
				return name ? `@${name}` : `@${userId}`;
			});
			let imgMatch;
			imageKeyRe.lastIndex = 0;
			while ((imgMatch = imageKeyRe.exec(parts[i])) !== null) if (imgMatch[2]) resources.push({
				type: "image",
				fileKey: imgMatch[2]
			});
		}
	}
	return parts.join("```");
}
function renderElement(el, ctx, resources) {
	switch (el.tag) {
		case "text": return applyStyle(el.text ?? "", el.style);
		case "a": {
			const label = el.text ?? el.href ?? "";
			return el.href ? `[${label}](${el.href})` : label;
		}
		case "at": {
			const userId = el.user_id ?? "";
			if (userId === "all" || userId === "all_members") return "@all";
			const info = ctx.mentionsByOpenId.get(userId);
			if (info) return info.key;
			return el.user_name ? `@${el.user_name}` : `@${userId}`;
		}
		case "img":
			if (el.image_key) {
				resources.push({
					type: "image",
					fileKey: el.image_key
				});
				return `![image](${el.image_key})`;
			}
			return "";
		case "media":
			if (el.file_key) {
				resources.push({
					type: "file",
					fileKey: el.file_key
				});
				return `<file key="${el.file_key}"/>`;
			}
			return "";
		case "code_block": return `\n\`\`\`${el.language ?? ""}\n${el.text ?? ""}\n\`\`\`\n`;
		case "hr": return "\n---\n";
		case "md": return processMdText(el.text ?? "", resources);
		default: return el.text ?? "";
	}
}
//#endregion
//#region src/normalize/converters/share.ts
const convertShareChat = async (raw, _ctx) => {
	return {
		content: `<group_card id="${safeParse(raw)?.chat_id ?? ""}"/>`,
		resources: []
	};
};
const convertShareUser = async (raw, _ctx) => {
	return {
		content: `<contact_card id="${safeParse(raw)?.user_id ?? ""}"/>`,
		resources: []
	};
};
//#endregion
//#region src/normalize/converters/sticker.ts
const convertSticker = async (raw, _ctx) => {
	const fileKey = safeParse(raw)?.file_key;
	if (!fileKey) return {
		content: "[sticker]",
		resources: []
	};
	const resources = [{
		type: "sticker",
		fileKey
	}];
	return {
		content: `<sticker key="${fileKey}"/>`,
		resources
	};
};
//#endregion
//#region src/normalize/converters/system.ts
const convertSystem = async (raw, _ctx) => {
	const parsed = safeParse(raw);
	if (!parsed || !parsed.template) return {
		content: "[system message]",
		resources: []
	};
	return {
		content: parsed.template.replace(/\{([a-z_]+)\}/g, (match, name) => {
			const val = parsed[name];
			if (Array.isArray(val)) return val.join(", ");
			if (typeof val === "string") return val;
			if (val == null) return "";
			return match;
		}).trim() || "[system message]",
		resources: []
	};
};
//#endregion
//#region src/normalize/converters/text.ts
const convertText = async (raw, _ctx) => {
	return {
		content: safeParse(raw)?.text ?? "",
		resources: []
	};
};
//#endregion
//#region src/normalize/converters/todo.ts
const convertTodo = async (raw, _ctx) => {
	const parsed = safeParse(raw);
	if (!parsed?.summary) return {
		content: "<todo>\n[todo]\n</todo>",
		resources: []
	};
	const lines = [];
	if (parsed.summary.title) lines.push(parsed.summary.title);
	const bodyText = extractPostPlainText(parsed.summary.content);
	if (bodyText) lines.push(bodyText);
	const due = millisToDatetime(parsed.due_time);
	if (due) lines.push(`Due: ${due}`);
	if (lines.length === 0) return {
		content: "<todo>\n[todo]\n</todo>",
		resources: []
	};
	return {
		content: `<todo>\n${lines.join("\n")}\n</todo>`,
		resources: []
	};
};
function extractPostPlainText(blocks) {
	if (!blocks) return "";
	const lines = [];
	for (const paragraph of blocks) {
		if (!Array.isArray(paragraph)) continue;
		const parts = [];
		for (const el of paragraph) if (el.tag === "text" && el.text) parts.push(el.text);
		else if (el.tag === "a" && el.text) parts.push(el.text);
		if (parts.length > 0) lines.push(parts.join(""));
	}
	return lines.join("\n");
}
//#endregion
//#region src/normalize/converters/video.ts
const convertVideo = async (raw, _ctx) => {
	const parsed = safeParse(raw);
	const fileKey = parsed?.file_key;
	if (!fileKey) return {
		content: "[video]",
		resources: []
	};
	const nameAttr = parsed?.file_name ? ` name="${escapeAttr(parsed.file_name)}"` : "";
	const durStr = formatDuration(parsed?.duration);
	return {
		content: `<video key="${fileKey}"${nameAttr}${durStr ? ` duration="${durStr}"` : ""}/>`,
		resources: [{
			type: "video",
			fileKey,
			fileName: parsed?.file_name,
			durationMs: parsed?.duration,
			coverImageKey: parsed?.image_key
		}]
	};
};
//#endregion
//#region src/normalize/converters/video-chat.ts
const convertVideoChat = async (raw, _ctx) => {
	const parsed = safeParse(raw);
	if (!parsed) return {
		content: "<meeting>\n[video chat]\n</meeting>",
		resources: []
	};
	const lines = [];
	if (parsed.topic) lines.push(`📹 ${parsed.topic}`);
	if (parsed.meet_number) lines.push(`🔢 ${parsed.meet_number}`);
	const start = millisToDatetime(parsed.start_time);
	if (start) lines.push(`🕙 ${start}`);
	return {
		content: `<meeting>\n${lines.length > 0 ? lines.join("\n") : "[video chat]"}\n</meeting>`,
		resources: []
	};
};
//#endregion
//#region src/normalize/converters/vote.ts
const convertVote = async (raw, _ctx) => {
	const parsed = safeParse(raw);
	if (!parsed || !parsed.topic && !parsed.options?.length) return {
		content: "<vote>\n[vote]\n</vote>",
		resources: []
	};
	const lines = [];
	if (parsed.topic) lines.push(parsed.topic);
	for (const opt of parsed.options ?? []) lines.push(`• ${opt}`);
	return {
		content: `<vote>\n${lines.join("\n")}\n</vote>`,
		resources: []
	};
};
//#endregion
//#region src/normalize/registry.ts
const converters = new Map([
	["text", convertText],
	["post", convertPost],
	["image", convertImage],
	["file", convertFile],
	["audio", convertAudio],
	["video", convertVideo],
	["media", convertVideo],
	["sticker", convertSticker],
	["interactive", convertInteractive],
	["merge_forward", convertMergeForward],
	["share_chat", convertShareChat],
	["share_user", convertShareUser],
	["location", convertLocation],
	["system", convertSystem],
	["vote", convertVote],
	["todo", convertTodo],
	["calendar", convertCalendar],
	["general_calendar", convertGeneralCalendar],
	["share_calendar_event", convertShareCalendarEvent],
	["folder", convertFolder],
	["hongbao", convertHongbao],
	["video_chat", convertVideoChat]
]);
/**
* Dispatch a message content to the matching converter, with uniform error
* containment — any thrown error is trapped and the fallback converter is
* invoked instead so that normalization never fails catastrophically.
*/
async function dispatchConvert(raw, msgType, ctx) {
	const fn = converters.get(msgType) ?? convertUnknown;
	try {
		return await fn(raw, ctx);
	} catch {
		return convertUnknown(raw, ctx);
	}
}
//#endregion
//#region src/normalize/bot-added.ts
function normalizeBotAdded(event, opts) {
	const chatId = event.chat_id;
	const operatorOpenId = event.operator_id?.open_id;
	if (!chatId || !operatorOpenId) return null;
	const botName = event.name ?? event.i18n_names?.zh_cn ?? event.i18n_names?.en_us ?? event.i18n_names?.ja_jp;
	return {
		chatId,
		operator: {
			openId: operatorOpenId,
			userId: event.operator_id?.user_id ?? void 0
		},
		botName,
		external: event.external,
		raw: opts?.includeRaw ? event : void 0
	};
}
//#endregion
//#region src/normalize/card-action.ts
function normalizeCardAction(event, opts) {
	const messageId = event.context?.open_message_id ?? event.open_message_id;
	const chatId = event.context?.open_chat_id ?? event.open_chat_id;
	const operatorOpenId = event.operator?.open_id;
	if (!messageId || !chatId || !operatorOpenId) return null;
	return {
		messageId,
		chatId,
		operator: {
			openId: operatorOpenId,
			userId: event.operator?.user_id,
			name: event.operator?.name
		},
		action: {
			value: event.action?.value,
			tag: event.action?.tag ?? "unknown",
			name: event.action?.name,
			option: event.action?.option,
			formValue: event.action?.form_value
		},
		raw: opts?.includeRaw ? event : void 0
	};
}
//#endregion
//#region src/normalize/comment.ts
function normalizeComment(event, opts) {
	const fileToken = event.file_token ?? event.notice_meta?.file_token;
	const fileType = event.file_type ?? event.notice_meta?.file_type;
	const commentId = event.comment_id;
	const userId = event.notice_meta?.from_user_id ?? event.user_id;
	const operatorOpenId = userId?.open_id;
	if (!fileToken || !fileType || !commentId || !operatorOpenId) return null;
	const tsStr = event.create_time ?? event.notice_meta?.timestamp ?? event.action_time;
	const timestamp = tsStr ? parseInt(tsStr, 10) : Date.now();
	return {
		fileToken,
		fileType,
		commentId,
		replyId: event.reply_id,
		operator: {
			openId: operatorOpenId,
			userId: userId?.user_id ?? void 0,
			unionId: userId?.union_id
		},
		mentionedBot: Boolean(event.is_mentioned ?? event.notice_meta?.is_mentioned ?? event.is_mention),
		timestamp: Number.isFinite(timestamp) ? timestamp : Date.now(),
		raw: opts?.includeRaw ? event : void 0
	};
}
//#endregion
//#region src/normalize/reaction.ts
function normalizeReaction(event, action, opts) {
	const messageId = event.message_id;
	const emojiType = event.reaction_type?.emoji_type;
	const operatorOpenId = event.user_id?.open_id;
	if (!messageId || !emojiType || !operatorOpenId) return null;
	const actionTimeStr = event.action_time;
	const actionTime = actionTimeStr ? parseInt(actionTimeStr, 10) : void 0;
	return {
		messageId,
		operator: {
			openId: operatorOpenId,
			userId: event.user_id?.user_id ?? void 0
		},
		emojiType,
		action,
		actionTime: actionTime != null && Number.isFinite(actionTime) ? actionTime : void 0,
		raw: opts?.includeRaw ? event : void 0
	};
}
//#endregion
//#region src/normalize/index.ts
/**
* Normalize a raw Feishu message event into a NormalizedMessage.
*
* Pipeline:
*   1. Extract mentions → build key/openId maps + bot detection
*   2. For `interactive` type, fetch full v2 card content if capability available
*   3. Build ConvertContext with injected capabilities
*   4. Dispatch to the matching converter (uniform error containment inside)
*   5. Run resolveMentions second pass — replace placeholders with @name
*   6. Assemble and return NormalizedMessage
*/
async function normalize(event, opts) {
	const msg = event.message;
	const botOpenId = opts.botIdentity?.openId;
	const { mentions, mentionsByOpenId, mentionList, mentionAll: mentionAllFromRaw, mentionedBot } = extractMentions(msg.mentions, botOpenId);
	const mentionAll = mentionAllFromRaw || detectMentionAllInContent(msg.content);
	const ctx = {
		messageId: msg.message_id,
		botOpenId,
		mentions,
		mentionsByOpenId,
		stripBotMentions: opts.stripBotMentions ?? true,
		fetchSubMessages: opts.fetchSubMessages,
		resolveUserName: opts.resolveUserName,
		batchResolveNames: opts.batchResolveNames,
		dispatch: dispatchConvert
	};
	const { content: rawContent, resources } = await dispatchConvert(msg.content, msg.message_type, ctx);
	const content = resolveMentions(rawContent, ctx);
	const senderOpenId = event.sender.sender_id.open_id;
	const senderFallbackId = event.sender.sender_id.user_id ?? event.sender.sender_id.union_id ?? "";
	const senderId = senderOpenId ?? senderFallbackId;
	const senderName = senderOpenId ? opts.resolveSenderName?.(senderOpenId) : void 0;
	const createMs = msg.create_time ? parseInt(msg.create_time, 10) : 0;
	return {
		messageId: msg.message_id,
		chatId: msg.chat_id,
		chatType: msg.chat_type,
		senderId,
		senderName,
		content,
		rawContentType: msg.message_type,
		resources,
		mentions: mentionList,
		mentionAll,
		mentionedBot,
		rootId: msg.root_id,
		threadId: msg.thread_id,
		replyToMessageId: msg.parent_id,
		createTime: Number.isFinite(createMs) ? createMs : 0,
		raw: opts.includeRaw ? event : void 0
	};
}
/**
* Detect `@_all` placeholder inside a raw Feishu content JSON string without
* parsing. We deliberately search the serialized form (not the parsed text),
* because the placeholder can appear in a `text` field (text/post) or inside
* nested content arrays (post). The placeholder is bounded by non-word chars
* on the right (whitespace, quote, punctuation) — on the left `@` is already
* a non-word char so no explicit boundary is needed.
*/
function detectMentionAllInContent(content) {
	if (!content) return false;
	return /@_all\b/.test(content);
}
//#endregion
//#region src/types.ts
var LarkChannelError = class extends Error {
	code;
	cause;
	context;
	constructor(code, message, opts) {
		super(message);
		this.name = "LarkChannelError";
		this.code = code;
		this.cause = opts?.cause;
		this.context = opts?.context;
	}
};
//#endregion
//#region src/outbound/errors.ts
/**
* Classify a raw error (typically from axios/fetch or a Feishu API response)
* into a LarkChannelError with a stable code.
*/
function classifyError(err, context) {
	if (err instanceof LarkChannelError) return err;
	return new LarkChannelError(inferCode(err), extractMessage(err), {
		cause: err,
		context
	});
}
function inferCode(err) {
	const raw = err;
	const status = raw?.response?.status ?? raw?.status;
	const feishuCode = raw?.response?.data?.code ?? raw?.data?.code ?? raw?.code;
	const msg = String(raw?.message ?? "").toLowerCase();
	if (typeof feishuCode === "number") {
		if (feishuCode === 230020 || feishuCode === 230017) return "target_revoked";
		if (feishuCode === 99991400 || feishuCode === 99991401) return "permission_denied";
		if (feishuCode === 230002 || feishuCode === 230001) return "format_error";
	}
	if (status === 429) return "rate_limited";
	if (status === 401 || status === 403) return "permission_denied";
	if (status === 400) return "format_error";
	if (status === 404) return "target_revoked";
	if (msg.startsWith("ssrf_blocked")) return "ssrf_blocked";
	if (msg.includes("timeout") || raw?.code === "ETIMEDOUT" || raw?.code === "ECONNABORTED") return "send_timeout";
	return "unknown";
}
function extractMessage(err) {
	const raw = err;
	return raw?.response?.data?.msg || raw?.response?.data?.message || raw?.message || String(err);
}
function isRetryable(err) {
	return err.code === "rate_limited" || err.code === "unknown";
}
function isFormatError(err) {
	return err.code === "format_error";
}
function isReplyTargetGone(err) {
	return err.code === "target_revoked";
}
//#endregion
//#region src/outbound/markdown/compose-mentions.ts
/**
* Build a text prefix that renders as real Feishu mentions when prepended
* to a text-type outbound message (the <at …> tag form).
*
* For post-type messages, the mentions should be injected as `at` elements
* at the beginning of the post body — use `composePostMentionElements`
* instead.
*/
function composeMentionsTextPrefix(mentions) {
	if (!mentions?.length) return "";
	const parts = [];
	for (const m of mentions) {
		if (!m.openId) continue;
		const name = m.name ?? "";
		parts.push(`<at user_id="${m.openId}">${name}</at>`);
	}
	return parts.length > 0 ? parts.join(" ") + " " : "";
}
//#endregion
//#region src/outbound/markdown/splitter.ts
/**
* Split a long markdown string into chunks under `limit` characters.
*
* Preserves code block integrity — if a chunk boundary would fall inside a
* fenced code block, the current chunk closes the fence and the next chunk
* reopens it with the same language tag.
*
* Prefers breaking before headings when possible.
*/
function splitWithCodeFences(text, limit) {
	if (text.length <= limit) return [text];
	const lines = text.split("\n");
	const out = [];
	let buf = [];
	let bufLen = 0;
	let fenceLang = null;
	const flush = () => {
		if (buf.length === 0) return;
		let chunk = buf.join("\n");
		if (fenceLang !== null) chunk += "\n```";
		out.push(chunk);
		buf = [];
		bufLen = 0;
		if (fenceLang !== null) {
			buf.push("```" + fenceLang);
			bufLen = buf[0].length;
		}
	};
	for (const line of lines) {
		const m = line.match(/^```(\w*)$/);
		const lineLen = line.length + (buf.length > 0 ? 1 : 0);
		const isHeading = /^#{1,6}\s/.test(line);
		const nearFull = bufLen > limit * .75;
		if (bufLen + lineLen > limit || isHeading && nearFull && buf.length > 0) flush();
		buf.push(line);
		bufLen += lineLen;
		if (m) fenceLang = fenceLang === null ? m[1] || "" : null;
	}
	flush();
	return out;
}
//#endregion
//#region src/outbound/markdown/optimize-style.ts
/**
* Pre-processing for Markdown text rendered through Feishu's native `tag: 'md'`
* post element. Without this, AI-produced markdown looks visually crammed —
* H1 headings render at huge size, code blocks butt against surrounding text,
* tables have no breathing room.
*
* Ported from openclaw-lark / src/card/markdown-style.ts. The two cardVersion
* branches preserve the original behavior:
*   - cardVersion=1 (Feishu post): heading demotion + blank-line compression
*   - cardVersion=2 (CardKit v2):  same plus <br> insertion around tables /
*     code blocks / consecutive headings
*
* channel currently only sends posts (msg_type='post'), so callers should
* pass cardVersion=1. The v2 path is kept for parity in case the channel
* grows native CardKit support later.
*/
function optimizeMarkdownStyle(text, cardVersion = 1) {
	try {
		return _optimize(text, cardVersion);
	} catch {
		return text;
	}
}
function _optimize(text, cardVersion) {
	const MARK = "___CB_";
	const codeBlocks = [];
	let r = text.replace(/(^|\n)(`{3,})([^\n]*)\n[\s\S]*?\n\2(?=\n|$)/g, (m, prefix = "") => {
		const block = m.slice(String(prefix).length);
		return `${prefix}${MARK}${codeBlocks.push(block) - 1}___`;
	});
	if (/^#{1,3} /m.test(text)) {
		r = r.replace(/^#{2,6} (.+)$/gm, "##### $1");
		r = r.replace(/^# (.+)$/gm, "#### $1");
	}
	if (cardVersion >= 2) {
		r = r.replace(/^(#{4,5} .+)\n{1,2}(#{4,5} )/gm, "$1\n<br>\n$2");
		r = r.replace(/^([^|\n].*)\n(\|.+\|)/gm, "$1\n\n$2");
		r = r.replace(/\n\n((?:\|.+\|[^\S\n]*\n?)+)/g, "\n\n<br>\n\n$1");
		r = r.replace(/((?:^\|.+\|[^\S\n]*\n?)+)/gm, (m, _table, offset) => {
			const after = r.slice(offset + m.length).replace(/^\n+/, "");
			if (!after || /^(---|#{4,5} |\*\*)/.test(after)) return m;
			return m + "\n<br>\n";
		});
		r = r.replace(/^((?!#{4,5} )(?!\*\*).+)\n\n(<br>)\n\n(\|)/gm, "$1\n$2\n$3");
		r = r.replace(/^(\*\*.+)\n\n(<br>)\n\n(\|)/gm, "$1\n$2\n\n$3");
		r = r.replace(/(\|[^\n]*\n)\n(<br>\n)((?!#{4,5} )(?!\*\*))/gm, "$1$2$3");
		codeBlocks.forEach((block, i) => {
			r = r.replace(`${MARK}${i}___`, `\n<br>\n${block}\n<br>\n`);
		});
	} else codeBlocks.forEach((block, i) => {
		r = r.replace(`${MARK}${i}___`, block);
	});
	r = r.replace(/\n{3,}/g, "\n\n");
	return r;
}
//#endregion
//#region src/outbound/markdown/to-post.ts
/**
* Convert a Markdown string to Feishu post JSON.
*
* The post body is a single `tag: 'md'` element whose `text` is the original
* markdown — Feishu's native renderer handles bold / italic / code / links /
* headings / lists / blockquotes / `<at>` mentions and code fences. The
* markdown is run through `optimizeMarkdownStyle` first to fix common visual
* issues (oversized H1/H2, excess blank lines, etc.).
*
* Mentions are rendered as `<at user_id="ou_xxx">name</at>` tokens prepended
* to the markdown, which the `md` element renders inline.
*/
function markdownToPost(md, opts) {
	const text = optimizeMarkdownStyle(composeMentionsTextPrefix(opts?.mentions ?? []) + md, 1);
	return { zh_cn: {
		title: opts?.title ?? "",
		content: [[{
			tag: "md",
			text
		}]]
	} };
}
/**
* Extract the plain text body from a post JSON. Used as a fallback when a
* post send is rejected with a format error and we need to re-send as a
* plain text message.
*/
function postToPlainText(post) {
	const body = post?.zh_cn;
	if (!body?.content) return "";
	const lines = [];
	for (const paragraph of body.content) {
		if (!Array.isArray(paragraph)) continue;
		const parts = [];
		for (const el of paragraph) switch (el.tag) {
			case "md":
			case "text":
			case "a":
				parts.push(el.text ?? "");
				break;
			case "at":
				parts.push(el.user_name ? `@${el.user_name}` : "");
				break;
			case "img":
				parts.push(el.image_key ? `[image]` : "");
				break;
		}
		lines.push(parts.join(""));
	}
	return lines.join("\n").trim();
}
//#endregion
//#region src/outbound/media/duration-mp4.ts
/**
* Parse MP4 video duration (ms) from a buffer by walking the ISO BMFF box
* hierarchy and finding `moov → mvhd`.
*
* The `mvhd` box version determines field widths:
* - version 0: creation/modification/timescale/duration = 4 bytes each
* - version 1: creation/modification = 8 bytes, timescale = 4 bytes,
*              duration = 8 bytes
*
* Returns undefined if the stream is not parseable or duration is unknown.
*/
function parseMp4Duration(buf) {
	if (!buf || buf.length < 16) return void 0;
	const moov = findBoxPayload(buf, 0, buf.length, "moov");
	if (!moov) return void 0;
	const mvhd = findBoxPayload(buf, moov.start, moov.end, "mvhd");
	if (!mvhd) return void 0;
	const p = mvhd.start;
	if (p + 4 > buf.length) return void 0;
	const version = buf.readUInt8(p);
	const base = p + 4;
	let timescale;
	let duration;
	if (version === 1) {
		if (base + 28 > buf.length) return void 0;
		timescale = buf.readUInt32BE(base + 16);
		duration = Number(buf.readBigUInt64BE(base + 20));
	} else {
		if (base + 16 > buf.length) return void 0;
		timescale = buf.readUInt32BE(base + 8);
		duration = buf.readUInt32BE(base + 12);
	}
	if (!timescale || !Number.isFinite(timescale) || !Number.isFinite(duration)) return;
	return Math.round(duration / timescale * 1e3);
}
/**
* Walk the boxes from [begin, end) looking for one whose type matches
* `name`. Returns the payload range (excluding the 8-byte header).
*/
function findBoxPayload(buf, begin, end, name) {
	let p = begin;
	while (p + 8 <= end && p + 8 <= buf.length) {
		const size = buf.readUInt32BE(p);
		const type = buf.slice(p + 4, p + 8).toString("ascii");
		const boxEnd = size === 1 ? p + Number(buf.readBigUInt64BE(p + 8)) : size === 0 ? end : p + size;
		if (boxEnd <= p || boxEnd > end) return void 0;
		if (type === name) return {
			start: size === 1 ? p + 16 : p + 8,
			end: boxEnd
		};
		p = boxEnd;
	}
}
//#endregion
//#region src/outbound/media/duration-ogg.ts
/**
* Parse Opus/OGG audio duration (ms) from a buffer.
*
* Scans backward for the last "OggS" page capture pattern (0x4f676753) and
* reads its granule_position (64-bit LE at offset +6 from magic). For Opus,
* granule_position is the number of decoded samples at 48 kHz — divide by
* 48 (samples per ms) to get milliseconds.
*
* Returns undefined if the buffer doesn't contain a valid Ogg stream or if
* the granule position is obviously invalid.
*/
function parseOpusDuration(buf) {
	if (!buf || buf.length < 27) return void 0;
	for (let i = buf.length - 27; i >= 0; i--) if (buf[i] === 79 && buf[i + 1] === 103 && buf[i + 2] === 103 && buf[i + 3] === 83) {
		const granule = buf.readBigInt64LE(i + 6);
		if (granule < BigInt(0)) return void 0;
		const ms = Number(granule) / 48;
		if (!Number.isFinite(ms) || ms < 0) return void 0;
		return Math.round(ms);
	}
}
//#endregion
//#region src/outbound/media/ssrf-guard.ts
/**
* CIDR blocks that must not be reached from a URL fetch triggered by the
* SDK. Covers loopback, link-local, private ranges, CGNAT, documentation.
*/
const BLOCKED_V4 = [
	[0, 8],
	[167772160, 8],
	[2130706432, 8],
	[2851995648, 16],
	[2886729728, 12],
	[3232235520, 16],
	[1681915904, 10],
	[3221225472, 24],
	[3221225984, 24],
	[3323068416, 15],
	[3325256704, 24],
	[3405803776, 24],
	[3758096384, 4],
	[4026531840, 4]
];
/**
* Validate that `url` does not resolve to an internal / reserved IP, and
* return the single IP that downstream fetching should pin to. Pinning is
* essential to close the DNS-rebinding TOCTOU gap: the attacker's DNS
* server could otherwise return a public IP to this check and a private
* IP (e.g. cloud metadata `169.254.169.254`) to the subsequent fetch.
*
* Allowlisted hostnames skip the public-IP check but still pin — pinning
* is only about consistency between check-time and use-time, and is
* harmless when the caller trusts the host.
*/
async function assertPublicUrl(url, opts = {}) {
	const u = new URL(url);
	if (u.protocol !== "http:" && u.protocol !== "https:") throw new Error(`ssrf_blocked: protocol ${u.protocol}`);
	const rawHost = u.hostname;
	const host = rawHost.startsWith("[") && rawHost.endsWith("]") ? rawHost.slice(1, -1) : rawHost;
	const allowlisted = opts.allowlist?.includes(host) ?? false;
	let resolvedIp;
	if ((0, net.isIP)(host)) {
		resolvedIp = host;
		if (!allowlisted) assertIpPublic(resolvedIp);
	} else {
		const records = await dns.promises.lookup(host, { all: true });
		if (records.length === 0) throw new Error(`ssrf_blocked: no DNS records for ${host}`);
		if (!allowlisted) for (const r of records) assertIpPublic(r.address);
		resolvedIp = records[0].address;
	}
	return {
		resolvedIp,
		originalHost: host
	};
}
function assertIpPublic(ip) {
	const v = (0, net.isIP)(ip);
	if (v === 4 && ipv4Blocked(ip)) throw new Error(`ssrf_blocked: ${ip}`);
	if (v === 6 && ipv6Blocked(ip)) throw new Error(`ssrf_blocked: ${ip}`);
	if (v === 0) throw new Error(`ssrf_blocked: not a valid IP: ${ip}`);
}
function ipv4Blocked(ip) {
	const parts = ip.split(".").map(Number);
	if (parts.length !== 4 || parts.some((p) => p < 0 || p > 255)) return true;
	const n = (parts[0] << 24 | parts[1] << 16 | parts[2] << 8 | parts[3]) >>> 0;
	return BLOCKED_V4.some(([net$1, bits]) => {
		const mask = bits === 0 ? 0 : -1 << 32 - bits >>> 0;
		return (n & mask) === (net$1 & mask);
	});
}
/**
* Hardcoded IPv6 CIDR blocks rejected outright (not delegated to v4).
* Each tuple is [network address as 128-bit bigint, prefix length].
*
* Note: `BigInt('0x...')` is used (instead of `0x...n` literals) so the
* code compiles under TypeScript `target: es6` — bigint literals need
* es2020. Semantically identical.
*/
const BLOCKED_V6 = [
	[BigInt("0xfe800000000000000000000000000000"), 10],
	[BigInt("0xfc000000000000000000000000000000"), 7],
	[BigInt("0xff000000000000000000000000000000"), 8],
	[BigInt("0x01000000000000000000000000000000"), 64],
	[BigInt("0x20010db8000000000000000000000000"), 32],
	[BigInt("0x20010000000000000000000000000000"), 32]
];
const HIGH96_MASK = BigInt("0xffffffffffffffffffffffff00000000");
const PREFIX_V4_MAPPED = BigInt("0x00000000000000000000ffff00000000");
const PREFIX_V4_COMPAT = BigInt(0);
const PREFIX_NAT64 = BigInt("0x0064ff9b000000000000000000000000");
const LOW32_MASK = BigInt("0xffffffff");
function ipv6Blocked(ip) {
	let n;
	try {
		n = parseIPv6(ip);
	} catch {
		return true;
	}
	const high96 = n & HIGH96_MASK;
	if (high96 === PREFIX_V4_MAPPED || high96 === PREFIX_NAT64 || high96 === PREFIX_V4_COMPAT) {
		const v4 = Number(n & LOW32_MASK);
		return ipv4Blocked([
			v4 >>> 24 & 255,
			v4 >>> 16 & 255,
			v4 >>> 8 & 255,
			v4 & 255
		].join("."));
	}
	return BLOCKED_V6.some(([net$2, bits]) => {
		const shift = BigInt(128 - bits);
		const mask = (BigInt(1) << BigInt(128)) - BigInt(1) >> shift << shift;
		return (n & mask) === (net$2 & mask);
	});
}
/**
* Parse any representation of an IPv6 address into a 128-bit bigint.
* Handles:
*   - `::` compression (`::1`, `fe80::1`, `::`)
*   - fully expanded form (`0:0:0:0:0:0:0:1`)
*   - IPv4-in-IPv6 mixed notation (`::ffff:127.0.0.1`, `64:ff9b::10.0.0.1`)
*   - Zone ID suffix (`fe80::1%eth0` — the suffix is ignored)
* Throws on any malformed input; the caller treats a throw as
* fail-closed.
*/
function parseIPv6(ip) {
	let addr = ip.split("%")[0].toLowerCase();
	const v4Suffix = addr.match(/:(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})$/);
	if (v4Suffix) {
		const parts = v4Suffix[1].split(".").map(Number);
		if (parts.length !== 4 || parts.some((p) => !Number.isFinite(p) || p < 0 || p > 255)) throw new Error("bad v4 suffix");
		const hex1 = (parts[0] << 8 | parts[1]).toString(16);
		const hex2 = (parts[2] << 8 | parts[3]).toString(16);
		addr = addr.slice(0, addr.length - v4Suffix[1].length) + `${hex1}:${hex2}`;
	}
	const halves = addr.split("::");
	if (halves.length > 2) throw new Error("multiple \"::\"");
	let groups;
	if (halves.length === 2) {
		const left = halves[0] ? halves[0].split(":") : [];
		const right = halves[1] ? halves[1].split(":") : [];
		const fill = 8 - left.length - right.length;
		if (fill < 0) throw new Error("too many groups");
		groups = [
			...left,
			...Array(fill).fill("0"),
			...right
		];
	} else groups = addr.split(":");
	if (groups.length !== 8) throw new Error("wrong group count");
	let n = BigInt(0);
	for (const g of groups) {
		if (!/^[0-9a-f]{0,4}$/.test(g)) throw new Error("bad group");
		n = n << BigInt(16) | BigInt(`0x${g || "0"}`);
	}
	return n;
}
//#endregion
//#region src/outbound/media/uploader.ts
/**
* POSIX directory prefixes that can never be a legitimate media source.
* Skipped on Windows (no equivalent single-prefix set exists there; a
* downstream allowlist is the right tool for Windows deployments).
*/
const POSIX_BLOCKED_PREFIXES = [
	"/etc/",
	"/proc/",
	"/sys/",
	"/dev/"
];
/** Max bytes accepted from a URL-based media source. Protects against
*  memory DoS when a malicious URL tries to return gigabytes. */
const URL_MAX_BYTES = 50 * 1024 * 1024;
var MediaUploader = class {
	client;
	config;
	constructor(client, config) {
		this.client = client;
		this.config = config;
	}
	async upload(input) {
		const buffer = await this.toBuffer(input.source);
		if (input.kind === "image") return this.uploadImage(buffer);
		if (input.kind === "audio") {
			const duration = this.resolveDuration(input, buffer);
			return this.uploadFile(buffer, "opus", input.fileName ?? "voice.opus", duration);
		}
		if (input.kind === "video") {
			const duration = this.resolveDuration(input, buffer);
			return this.uploadFile(buffer, "mp4", input.fileName ?? "video.mp4", duration);
		}
		return this.uploadFile(buffer, "stream", input.fileName ?? "upload.bin");
	}
	/**
	* Materialize `source` into a Buffer. A string is treated as:
	*   • `http://` / `https://` URL — fetch with SSRF guard
	*   • anything else — local filesystem path, read directly
	*/
	async toBuffer(source) {
		if (Buffer.isBuffer(source)) return source;
		if (!/^https?:\/\//i.test(source)) {
			const allowedDirs = this.config?.allowedFileDirs;
			if (!allowedDirs || allowedDirs.length === 0) throw new LarkChannelError("upload_failed", "local file source requires `outbound.allowedFileDirs` to be configured");
			const resolved = path.default.resolve(source);
			this.assertNotOnPosixBlocklist(resolved);
			try {
				const realPath = await fs.default.promises.realpath(resolved);
				this.assertNotOnPosixBlocklist(realPath);
				if (!(await Promise.all(allowedDirs.map(async (d) => {
					const r = path.default.resolve(d);
					try {
						return await fs.default.promises.realpath(r);
					} catch {
						return r;
					}
				}))).some((d) => realPath === d || realPath.startsWith(d + path.default.sep))) throw new LarkChannelError("upload_failed", `file path is outside allowed directories: ${realPath}`);
				const fh = await fs.default.promises.open(realPath, "r");
				try {
					if (!(await fh.stat()).isFile()) throw new LarkChannelError("upload_failed", `not a regular file: ${realPath}`);
					return await fh.readFile();
				} finally {
					await fh.close();
				}
			} catch (e) {
				if (e instanceof LarkChannelError) throw e;
				throw new LarkChannelError("upload_failed", `source is neither an http(s) URL nor a readable local file: ${source}`, { cause: e });
			}
		}
		const ssrf = this.config?.ssrfGuard;
		const guardEnabled = ssrf !== false;
		let resolvedIp;
		if (guardEnabled) {
			const ssrfOpts = typeof ssrf === "object" && ssrf ? { allowlist: ssrf.allowlist } : {};
			try {
				({resolvedIp} = await assertPublicUrl(source, ssrfOpts));
			} catch (e) {
				throw new LarkChannelError("ssrf_blocked", `URL blocked: ${String(e)}`, { cause: e });
			}
		}
		try {
			const requestOpts = {
				url: source,
				method: "GET",
				responseType: "arraybuffer",
				timeout: 15e3,
				maxContentLength: URL_MAX_BYTES,
				maxBodyLength: URL_MAX_BYTES,
				maxRedirects: 0
			};
			if (resolvedIp) {
				const agent = makePinnedAgent(source, resolvedIp);
				requestOpts.httpAgent = agent;
				requestOpts.httpsAgent = agent;
			}
			const res = await this.client.httpInstance.request(requestOpts);
			return Buffer.from(res);
		} catch (e) {
			throw new LarkChannelError("upload_failed", `fetch source URL failed`, { cause: e });
		}
	}
	/**
	* Reject paths pointing at POSIX system directories (`/etc`, `/proc`,
	* `/sys`, `/dev`). Called twice by the caller — once on the resolved
	* path pre-I/O, again on the realpath — so both direct hits
	* (`/etc/passwd`) and macOS alias hits (`/etc` → `/private/etc`) are
	* caught.
	*/
	assertNotOnPosixBlocklist(p) {
		if (process.platform === "win32") return;
		if (POSIX_BLOCKED_PREFIXES.some((pre) => p === pre.slice(0, -1) || p.startsWith(pre))) throw new LarkChannelError("upload_failed", `file path is not allowed: ${p}`);
	}
	resolveDuration(input, buffer) {
		if (input.duration != null && input.duration > 0) return input.duration;
		const parsed = input.kind === "audio" ? parseOpusDuration(buffer) : input.kind === "video" ? parseMp4Duration(buffer) : void 0;
		if (parsed != null) return parsed;
		throw new LarkChannelError("upload_failed", `duration could not be determined for ${input.kind}; pass it explicitly`);
	}
	async uploadImage(buffer) {
		try {
			const r = await this.client.im.v1.image.create({ data: {
				image_type: "message",
				image: buffer
			} });
			const key = r?.image_key ?? r?.data?.image_key;
			if (!key) throw new Error("image_key missing in upload response");
			return {
				kind: "image",
				fileKey: key
			};
		} catch (e) {
			throw new LarkChannelError("upload_failed", `image upload failed`, { cause: e });
		}
	}
	async uploadFile(buffer, fileType, fileName, durationMs) {
		try {
			const data = {
				file_type: fileType,
				file_name: fileName,
				file: buffer
			};
			if (durationMs != null) data.duration = durationMs;
			const r = await this.client.im.v1.file.create({ data });
			const key = r?.file_key ?? r?.data?.file_key;
			if (!key) throw new Error("file_key missing in upload response");
			return {
				kind: fileType === "opus" ? "audio" : fileType === "mp4" ? "video" : "file",
				fileKey: key,
				durationMs
			};
		} catch (e) {
			if (e instanceof LarkChannelError) throw e;
			throw new LarkChannelError("upload_failed", `file upload failed`, { cause: e });
		}
	}
};
/**
* Build a per-request http(s) Agent whose DNS lookup always returns
* `pinnedIp`, regardless of what hostname Node would otherwise resolve.
* The URL's original hostname is preserved on the wire, which keeps TLS
* SNI and certificate verification working. This closes the window where
* a malicious DNS server could return a different (private) IP on the
* second resolution triggered by the actual fetch.
*/
function makePinnedAgent(url, pinnedIp) {
	const agent = new (url.startsWith("https:") ? https.default.Agent : http.default.Agent)();
	const family = pinnedIp.includes(":") ? 6 : 4;
	const lookup = (_hostname, _opts, cb) => {
		cb(null, pinnedIp, family);
	};
	const origCreateConnection = agent.createConnection.bind(agent);
	agent.createConnection = (opts, cb) => origCreateConnection({
		...opts,
		lookup
	}, cb);
	return agent;
}
//#endregion
//#region src/outbound/retry.ts
/**
* Execute `op` with exponential backoff. Only retries errors classified as
* retryable (rate_limited / unknown). Business errors (format / revoked /
* permission / timeout) fail fast and bubble up.
*/
async function retry(op, opts = {}) {
	const max = opts.maxAttempts ?? 3;
	const base = opts.baseDelayMs ?? 500;
	let lastErr;
	for (let attempt = 1; attempt <= max; attempt++) try {
		return await op(attempt);
	} catch (raw) {
		const err = classifyError(raw, { attempt });
		lastErr = err;
		if (attempt >= max || !isRetryable(err)) throw err;
		await sleep(base * 3 ** (attempt - 1));
	}
	throw lastErr;
}
function sleep(ms) {
	return new Promise((r) => setTimeout(r, ms));
}
//#endregion
//#region src/outbound/routing.ts
/**
* Infer Feishu's `receive_id_type` from the prefix of a target id.
*
*   oc_*          → chat_id
*   ou_*          → open_id
*   on_*          → union_id
*   contains '@'  → email
*   fallback      → user_id
*/
function detectReceiveIdType(to) {
	if (!to) throw new Error("empty receive_id");
	if (to.startsWith("oc_")) return "chat_id";
	if (to.startsWith("ou_")) return "open_id";
	if (to.startsWith("on_")) return "union_id";
	if (to.includes("@")) return "email";
	return "user_id";
}
//#endregion
//#region src/outbound/streaming/throttle.ts
/**
* Dual-threshold throttle: fires `flush` either when `ms` have elapsed
* since the last fire OR when `chars` characters have accumulated.
*
* Usage inside a stream controller:
*   const t = new Throttle({ ms: 100, chars: 50 }, () => doPatch(buffer));
*   await t.note(deltaLen);       // may or may not fire, schedule the rest
*   await t.flushNow();           // end-of-stream force flush
*/
var Throttle = class {
	opts;
	fire;
	pendingChars = 0;
	timer;
	/** Tracks the in-flight fire as a Promise so `flushNow` can await it. */
	inFlight;
	lastFireAt = 0;
	constructor(opts, fire) {
		this.opts = opts;
		this.fire = fire;
	}
	/**
	* Accumulate bytes and decide whether to fire now, schedule a timer,
	* or do nothing (a fire is already scheduled).
	*/
	note(deltaChars) {
		this.pendingChars += deltaChars;
		if (this.pendingChars >= this.opts.chars) {
			this.fireSoon(0);
			return;
		}
		if (!this.timer) {
			const elapsed = Date.now() - this.lastFireAt;
			const wait = Math.max(0, this.opts.ms - elapsed);
			this.fireSoon(wait);
		}
	}
	fireSoon(delay) {
		if (this.timer) clearTimeout(this.timer);
		this.timer = setTimeout(() => {
			this.timer = void 0;
			this.doFire();
		}, delay);
	}
	/**
	* Force-flush everything accumulated so far, including any content that
	* arrived during an in-flight fire. Waits for the in-flight fire to
	* complete, then fires once more to capture the final state — without
	* this second fire, the last chunk of appended content would be missed
	* (the in-flight fire captured a snapshot BEFORE those chunks arrived).
	*/
	async flushNow() {
		if (this.timer) {
			clearTimeout(this.timer);
			this.timer = void 0;
		}
		if (this.inFlight) await this.inFlight;
		await this.doFire();
	}
	async doFire() {
		if (this.inFlight) {
			this.fireSoon(this.opts.ms);
			return;
		}
		const p = (async () => {
			this.pendingChars = 0;
			this.lastFireAt = Date.now();
			await this.fire();
		})();
		this.inFlight = p;
		try {
			await p;
		} finally {
			this.inFlight = void 0;
		}
	}
	dispose() {
		if (this.timer) {
			clearTimeout(this.timer);
			this.timer = void 0;
		}
	}
};
//#endregion
//#region src/outbound/streaming/update-queue.ts
/**
* A per-stream FIFO queue that serializes async updates, so that concurrent
* calls to `append()` / `update()` always result in PATCH operations
* happening in submission order.
*/
var UpdateQueue = class {
	tail = Promise.resolve();
	enqueue(task) {
		const next = this.tail.then(task, task);
		this.tail = next.then(() => void 0, () => void 0);
		return next;
	}
	drain() {
		return this.tail;
	}
};
//#endregion
//#region src/outbound/streaming/card-stream.ts
const DEFAULT_THROTTLE_MS$1 = 100;
const DEFAULT_THROTTLE_CHARS$1 = 50;
/**
* Streaming card reply — the caller provides an initial card JSON and a
* producer that drives incremental `update()` calls with full or partial
* new card state.
*
* Updates are throttled + serialized. On producer exception, the last
* known state is kept and an error footer element is appended.
*/
var CardStreamControllerImpl = class {
	sender;
	to;
	idType;
	opts;
	_current;
	_messageId = "";
	throttle;
	queue = new UpdateQueue();
	constructor(sender, to, idType, opts, initial) {
		this.sender = sender;
		this.to = to;
		this.idType = idType;
		this.opts = opts;
		this._current = initial;
		const cfg = sender.config;
		this.throttle = new Throttle({
			ms: cfg.streamThrottleMs ?? DEFAULT_THROTTLE_MS$1,
			chars: cfg.streamThrottleChars ?? DEFAULT_THROTTLE_CHARS$1
		}, () => this.patch());
	}
	get messageId() {
		return this._messageId;
	}
	get current() {
		return this._current;
	}
	async update(next) {
		const nextCard = typeof next === "function" ? next(this._current) : next;
		this._current = nextCard;
		this.throttle.note(JSON.stringify(nextCard).length);
	}
	async run(producer) {
		await this.sendInitial();
		try {
			await producer(this);
		} catch (e) {
			await this.failTerminal(e);
			throw e;
		}
		await this.completeTerminal();
		return { messageId: this._messageId };
	}
	async sendInitial() {
		const id = await this.sender.sendOneWithFallback({
			to: this.to,
			idType: this.idType,
			msgType: "interactive",
			content: this._current,
			replyTo: this.opts.replyTo,
			replyInThread: this.opts.replyInThread
		});
		this._messageId = id;
	}
	async patch() {
		if (!this._messageId) return;
		const snapshot = this._current;
		await this.queue.enqueue(async () => {
			await this.sender.patchCard(this._messageId, snapshot);
		});
	}
	async completeTerminal() {
		await this.throttle.flushNow();
		await this.queue.drain();
	}
	async failTerminal(_err) {
		this.throttle.dispose();
		const withFooter = appendErrorFooter(this._current);
		await this.queue.enqueue(async () => {
			try {
				await this.sender.patchCard(this._messageId, withFooter);
			} catch {}
		});
		await this.queue.drain();
	}
};
function appendErrorFooter(card) {
	const cardObj = card;
	const elements = Array.isArray(cardObj?.elements) ? [...cardObj.elements] : [];
	elements.push({
		tag: "note",
		elements: [{
			tag: "plain_text",
			content: "⚠️ 生成中断"
		}]
	});
	return {
		...card,
		elements
	};
}
var CardStreamController = class {
	impl;
	constructor(sender, to, idType, opts, initial) {
		this.impl = new CardStreamControllerImpl(sender, to, idType, opts, initial);
	}
	run(producer) {
		return this.impl.run(producer);
	}
};
//#endregion
//#region src/outbound/streaming/markdown-stream.ts
const DEFAULT_THROTTLE_MS = 100;
const DEFAULT_THROTTLE_CHARS = 50;
const DEFAULT_INITIAL = "Thinking...";
const DEFAULT_EMPTY = "(no content)";
const INITIAL_SUMMARY = "[Generating...]";
const SUMMARY_MAX_CHARS = 50;
const ERROR_FOOTER = "\n\n— _(Generation interrupted)_";
const DEFAULT_MAX_ELEMENT_CHARS = 3e4;
const ELEMENT_ID = "stream_md";
/**
* Shorten a markdown content string into a single-line preview suitable for
* the card's `summary.content` — shown in chat lists / message previews.
*/
function truncateSummary(text, max = SUMMARY_MAX_CHARS) {
	if (!text) return "";
	const cleaned = text.replace(/\s+/g, " ").trim();
	return cleaned.length <= max ? cleaned : cleaned.slice(0, max - 1) + "…";
}
/**
* Build the initial card JSON for a streaming markdown reply.
*
* `streaming_mode: true` tells Feishu client to render incremental updates
* (via the cardElement.content API) as a native typewriter animation.
*
* `print_strategy: 'fast'` means: when a new full-content update arrives,
* immediately show any already-buffered-but-not-yet-animated text, so the
* display doesn't lag behind the upstream token rate.
*/
function buildStreamingCard(initialText) {
	return {
		schema: "2.0",
		config: {
			streaming_mode: true,
			summary: { content: INITIAL_SUMMARY },
			streaming_config: {
				print_frequency_ms: { default: 70 },
				print_step: { default: 1 },
				print_strategy: "fast"
			}
		},
		body: { elements: [{
			tag: "markdown",
			element_id: ELEMENT_ID,
			content: initialText
		}] }
	};
}
/**
* Streaming markdown reply that uses Feishu's native typewriter effect
* (cardkit.v1.cardElement.content).
*
* Flow:
*   1. ensureStarted(): create a card instance with streaming_mode=true,
*      send it as an interactive message referencing the card_id.
*   2. append(chunk) / setContent(full): accumulate locally, throttle.
*   3. throttle fires: updateCardElementContent(content, sequence++)
*      Feishu client renders typewriter animation based on the diff.
*   4. completeTerminal(): drain queue + finishStreamingCard to disable
*      streaming_mode (removes the typing cursor).
*   5. producer throws → append ERROR_FOOTER + finish stream → rethrow.
*
* Rollover: when the current card's element content exceeds
* `streamMaxElementChars` (Feishu rejects oversized elements with code
* 230099), the controller splits the content at a safe boundary, finalizes
* the current card with the head, and creates a fresh streaming card to
* continue with the tail. Multiple rollovers within one stream are
* supported. The head card's messageId is what `messageId` / `run()`
* return; follow-up cards are tracked internally.
*/
var MarkdownStreamControllerImpl = class {
	sender;
	to;
	idType;
	opts;
	/** Content of the current (latest) card's markdown element. */
	content = "";
	_messageId = "";
	cardId = "";
	sequence = 0;
	throttle;
	queue = new UpdateQueue();
	started = false;
	/**
	* Set when an unrecoverable update error occurs; subsequent pushes /
	* rollovers are skipped to avoid spamming the API. Terminal cleanup
	* still runs on a best-effort basis.
	*/
	streamingFailed = false;
	/** messageIds of every rollover card created after the head. */
	rolloverMessageIds = [];
	maxChars;
	constructor(sender, to, idType, opts) {
		this.sender = sender;
		this.to = to;
		this.idType = idType;
		this.opts = opts;
		const cfg = this.sender.config;
		this.throttle = new Throttle({
			ms: cfg.streamThrottleMs ?? DEFAULT_THROTTLE_MS,
			chars: cfg.streamThrottleChars ?? DEFAULT_THROTTLE_CHARS
		}, () => this.pushContent());
		this.maxChars = cfg.streamMaxElementChars ?? DEFAULT_MAX_ELEMENT_CHARS;
	}
	get messageId() {
		return this._messageId;
	}
	async append(chunk) {
		if (!chunk) return;
		await this.ensureStarted();
		this.content += chunk;
		this.throttle.note(chunk.length);
	}
	async setContent(full) {
		await this.ensureStarted();
		this.content = full ?? "";
		this.throttle.note(Number.MAX_SAFE_INTEGER);
	}
	async run(producer) {
		await this.ensureStarted();
		try {
			await producer(this);
		} catch (e) {
			await this.failTerminal(e);
			throw e;
		}
		await this.completeTerminal();
		return { messageId: this._messageId };
	}
	async ensureStarted() {
		if (this.started) return;
		this.started = true;
		const cardSpec = buildStreamingCard((this.sender.config.streamInitialText ?? DEFAULT_INITIAL) || "...");
		this.cardId = await this.sender.createCardInstance(cardSpec);
		this._messageId = await this.sender.sendCardByReference(this.to, this.idType, this.cardId, this.opts);
	}
	async pushContent() {
		if (!this.cardId || this.streamingFailed) return;
		await this.queue.enqueue(async () => {
			if (this.streamingFailed) return;
			try {
				await this.pushSnapshot();
			} catch (e) {
				this.streamingFailed = true;
				this.sender.logger.warn?.("[stream] update failed", e);
			}
		});
	}
	/**
	* Roll over until content fits and then PATCH the latest card. Caller
	* must already be inside a queue task — no enqueue here. Used by
	* pushContent and the terminal helpers so they share rollover behavior.
	*/
	async pushSnapshot() {
		while (this.content.length > this.maxChars) await this.rollover();
		const snapshot = this.content || "...";
		await this.sender.updateCardElementContent(this.cardId, ELEMENT_ID, snapshot, ++this.sequence);
	}
	/**
	* Finalize the current card with as much head content as fits and start
	* a fresh streaming card for the tail. Called from inside the queue task
	* so rollover steps are serialized with regular updates.
	*/
	async rollover() {
		const chunks = splitWithCodeFences(this.content, this.maxChars);
		if (chunks.length < 2) throw new Error("rollover: content not splittable below limit");
		const head = chunks[0];
		const tail = chunks.slice(1).join("\n");
		await this.sender.updateCardElementContent(this.cardId, ELEMENT_ID, head, ++this.sequence);
		try {
			await this.sender.finishStreamingCard(this.cardId, ++this.sequence, truncateSummary(head));
		} catch {}
		const cardSpec = buildStreamingCard(tail || "...");
		const newCardId = await this.sender.createCardInstance(cardSpec);
		const newMessageId = await this.sender.sendCardByReference(this.to, this.idType, newCardId, this.opts);
		this.cardId = newCardId;
		this.content = tail;
		this.sequence = 0;
		this.rolloverMessageIds.push(newMessageId);
	}
	async completeTerminal() {
		await this.throttle.flushNow();
		await this.queue.drain();
		if (!this.cardId) return;
		if (!this.content && !this.streamingFailed) {
			await this.queue.enqueue(async () => {
				try {
					await this.sender.updateCardElementContent(this.cardId, ELEMENT_ID, DEFAULT_EMPTY, ++this.sequence);
				} catch {}
			});
			await this.queue.drain();
		}
		try {
			await this.sender.finishStreamingCard(this.cardId, ++this.sequence, truncateSummary(this.content || DEFAULT_EMPTY));
		} catch (e) {
			this.sender.logger.warn?.("[stream] finishStreamingCard failed", e);
		}
	}
	async failTerminal(_err) {
		this.throttle.dispose();
		if (!this.cardId) return;
		this.content = (this.content || "") + ERROR_FOOTER;
		await this.queue.enqueue(async () => {
			try {
				await this.pushSnapshot();
			} catch {}
		});
		await this.queue.drain();
		try {
			await this.sender.finishStreamingCard(this.cardId, ++this.sequence, truncateSummary(this.content));
		} catch {}
	}
};
var MarkdownStreamController = class {
	impl;
	constructor(sender, to, idType, opts) {
		this.impl = new MarkdownStreamControllerImpl(sender, to, idType, opts);
	}
	run(producer) {
		return this.impl.run(producer);
	}
};
//#endregion
//#region src/outbound/sender.ts
const DEFAULT_CHUNK_LIMIT = 3500;
var OutboundSender = class {
	client;
	config;
	logger;
	uploader;
	chunkLimit;
	constructor(client, config, logger) {
		this.client = client;
		this.config = config;
		this.logger = logger;
		this.uploader = new MediaUploader(client, config);
		this.chunkLimit = config.textChunkLimit ?? DEFAULT_CHUNK_LIMIT;
	}
	async send(to, input, opts = {}) {
		const idType = detectReceiveIdType(to);
		if ("markdown" in input) return this.sendMarkdown(to, idType, input.markdown, opts);
		if ("text" in input) return this.sendText(to, idType, input.text, opts);
		if ("post" in input) return this.sendPost(to, idType, input.post, opts);
		if ("image" in input) return this.sendImage(to, idType, input.image, opts);
		if ("file" in input) return this.sendFile(to, idType, input.file, opts);
		if ("audio" in input) return this.sendAudio(to, idType, input.audio, opts);
		if ("video" in input) return this.sendVideo(to, idType, input.video, opts);
		if ("card" in input) return this.sendCard(to, idType, input.card, opts);
		if ("cardId" in input) {
			const id = await this.sendCardByReference(to, idType, input.cardId, opts);
			return this.makeResult([id]);
		}
		if ("shareChat" in input) return this.sendShareChat(to, idType, input.shareChat.chatId, opts);
		if ("shareUser" in input) return this.sendShareUser(to, idType, input.shareUser.userId, opts);
		if ("sticker" in input) return this.sendSticker(to, idType, input.sticker.fileKey, opts);
		throw new LarkChannelError("format_error", "unrecognized SendInput shape");
	}
	async stream(to, input, opts = {}) {
		const idType = detectReceiveIdType(to);
		if ("markdown" in input) return new MarkdownStreamController(this, to, idType, opts).run(input.markdown);
		if ("card" in input) return new CardStreamController(this, to, idType, opts, input.card.initial).run(input.card.producer);
		throw new LarkChannelError("format_error", "unrecognized StreamInput shape");
	}
	async sendMarkdown(to, idType, md, opts) {
		const chunks = splitWithCodeFences(md, this.chunkLimit);
		const ids = [];
		for (let i = 0; i < chunks.length; i++) {
			const post = this.convertMarkdown(chunks[i], i === 0 ? opts.mentions : void 0);
			const id = await this.sendOneWithFallback({
				to,
				idType,
				msgType: "post",
				content: post,
				replyTo: replyTargetForChunk(i, opts, ids),
				replyInThread: opts.replyInThread
			});
			ids.push(id);
		}
		return this.makeResult(ids);
	}
	async sendText(to, idType, text, opts) {
		const chunks = splitPlain(composeMentionsTextPrefix(opts.mentions ?? []) + text, this.chunkLimit);
		const ids = [];
		for (let i = 0; i < chunks.length; i++) {
			const id = await this.sendOneWithFallback({
				to,
				idType,
				msgType: "text",
				content: { text: chunks[i] },
				replyTo: replyTargetForChunk(i, opts, ids),
				replyInThread: opts.replyInThread
			});
			ids.push(id);
		}
		return this.makeResult(ids);
	}
	async sendPost(to, idType, post, opts) {
		return { messageId: await this.sendOneWithFallback({
			to,
			idType,
			msgType: "post",
			content: post,
			replyTo: opts.replyTo,
			replyInThread: opts.replyInThread
		}) };
	}
	convertMarkdown(md, mentions) {
		const conv = this.config.markdownConverter;
		if (typeof conv === "function") return conv(md);
		return markdownToPost(md, { mentions });
	}
	async sendImage(to, idType, input, opts) {
		const up = await this.uploader.upload({
			kind: "image",
			source: input.source
		});
		return { messageId: await this.sendOneWithFallback({
			to,
			idType,
			msgType: "image",
			content: { image_key: up.fileKey },
			replyTo: opts.replyTo,
			replyInThread: opts.replyInThread
		}) };
	}
	async sendFile(to, idType, input, opts) {
		const up = await this.uploader.upload({
			kind: "file",
			source: input.source,
			fileName: input.fileName
		});
		return { messageId: await this.sendOneWithFallback({
			to,
			idType,
			msgType: "file",
			content: { file_key: up.fileKey },
			replyTo: opts.replyTo,
			replyInThread: opts.replyInThread
		}) };
	}
	async sendAudio(to, idType, input, opts) {
		const up = await this.uploader.upload({
			kind: "audio",
			source: input.source,
			duration: input.duration
		});
		return { messageId: await this.sendOneWithFallback({
			to,
			idType,
			msgType: "audio",
			content: { file_key: up.fileKey },
			replyTo: opts.replyTo,
			replyInThread: opts.replyInThread
		}) };
	}
	async sendVideo(to, idType, input, opts) {
		const content = { file_key: (await this.uploader.upload({
			kind: "video",
			source: input.source,
			duration: input.duration
		})).fileKey };
		if (input.coverImageKey) content.image_key = input.coverImageKey;
		return { messageId: await this.sendOneWithFallback({
			to,
			idType,
			msgType: "media",
			content,
			replyTo: opts.replyTo,
			replyInThread: opts.replyInThread
		}) };
	}
	async sendCard(to, idType, card, opts) {
		return { messageId: await this.sendOneWithFallback({
			to,
			idType,
			msgType: "interactive",
			content: card,
			replyTo: opts.replyTo,
			replyInThread: opts.replyInThread
		}) };
	}
	async sendShareChat(to, idType, chatId, opts) {
		return { messageId: await this.sendOneWithFallback({
			to,
			idType,
			msgType: "share_chat",
			content: { chat_id: chatId },
			replyTo: opts.replyTo,
			replyInThread: opts.replyInThread
		}) };
	}
	async sendShareUser(to, idType, userId, opts) {
		return { messageId: await this.sendOneWithFallback({
			to,
			idType,
			msgType: "share_user",
			content: { user_id: userId },
			replyTo: opts.replyTo,
			replyInThread: opts.replyInThread
		}) };
	}
	async sendSticker(to, idType, fileKey, opts) {
		return { messageId: await this.sendOneWithFallback({
			to,
			idType,
			msgType: "sticker",
			content: { file_key: fileKey },
			replyTo: opts.replyTo,
			replyInThread: opts.replyInThread
		}) };
	}
	/**
	* Send once with fallback for format errors (post → text) and for
	* vanished reply targets (reply → create).
	*/
	async sendOneWithFallback(args) {
		try {
			return await this.rawSendWithRetry(args);
		} catch (e) {
			const err = classifyError(e, { to: args.to });
			if (isReplyTargetGone(err) && args.replyTo) return this.rawSendWithRetry({
				...args,
				replyTo: void 0
			});
			if (isFormatError(err) && args.msgType === "post") {
				const plain = postToPlainText(args.content);
				return this.rawSendWithRetry({
					...args,
					msgType: "text",
					content: { text: plain || "[message]" }
				});
			}
			throw err;
		}
	}
	async rawSendWithRetry(args) {
		return retry(() => this.rawSend(args), this.config.retry);
	}
	async rawSend(args) {
		const payload = {
			receive_id: args.to,
			msg_type: args.msgType,
			content: JSON.stringify(args.content)
		};
		if (args.replyTo) {
			const id = (await this.client.im.v1.message.reply({
				path: { message_id: args.replyTo },
				data: {
					content: payload.content,
					msg_type: args.msgType,
					reply_in_thread: args.replyInThread
				}
			})).data?.message_id;
			if (!id) throw new LarkChannelError("unknown", "message_id missing from reply response");
			return id;
		}
		const id = (await this.client.im.v1.message.create({
			params: { receive_id_type: args.idType },
			data: payload
		})).data?.message_id;
		if (!id) throw new LarkChannelError("unknown", "message_id missing from create response");
		return id;
	}
	/**
	* Full card replace via im.v1.message.patch — used by CardStreamController
	* and the public `channel.updateCard()` low-level API. Not suitable for
	* text streaming (use native cardkit streaming below for that).
	*/
	async patchCard(messageId, card) {
		await this.client.im.v1.message.patch({
			path: { message_id: messageId },
			data: { content: JSON.stringify(card) }
		});
	}
	/**
	* Create a card instance via cardkit.v1.card.create. Used to get a
	* `card_id` that can be referenced from messages AND updated with
	* native streaming APIs (cardElement.content for typewriter effect).
	*/
	async createCardInstance(spec) {
		const cardId = (await this.client.cardkit.v1.card.create({ data: {
			type: "card_json",
			data: JSON.stringify(spec)
		} })).data?.card_id;
		if (!cardId) throw new LarkChannelError("unknown", "cardkit.card.create returned no card_id");
		return cardId;
	}
	/**
	* Send an interactive message that references a pre-created card instance
	* by card_id. Returns the resulting message_id.
	*/
	async sendCardByReference(to, idType, cardId, opts) {
		return this.rawSendWithRetry({
			to,
			idType,
			msgType: "interactive",
			content: {
				type: "card",
				data: { card_id: cardId }
			},
			replyTo: opts.replyTo,
			replyInThread: opts.replyInThread
		});
	}
	/**
	* Full card replace by `card_id` via cardkit.v1.card.update — the entity-API
	* counterpart to {@link patchCard} (which targets a message_id). `sequence`
	* must increase monotonically per card so out-of-order updates are rejected
	* server-side rather than silently clobbering newer content. `uuid` is
	* derived from cardId + sequence for idempotency.
	*/
	async updateCardFull(cardId, cardJson, sequence) {
		await this.client.cardkit.v1.card.update({
			path: { card_id: cardId },
			data: {
				card: {
					type: "card_json",
					data: JSON.stringify(cardJson)
				},
				sequence,
				uuid: `u_${cardId}_${sequence}`
			}
		});
	}
	/**
	* Stream update: replace a card element's content and let Feishu render
	* the incremental diff as a typewriter animation. `sequence` must be
	* monotonically increasing per card (duplicates/out-of-order → rejected
	* server-side). `uuid` is required for idempotency — we compose it from
	* cardId + sequence so two concurrent calls with the same seq would
	* dedupe rather than both take effect.
	*/
	async updateCardElementContent(cardId, elementId, content, sequence) {
		await this.client.cardkit.v1.cardElement.content({
			path: {
				card_id: cardId,
				element_id: elementId
			},
			data: {
				content,
				sequence,
				uuid: `c_${cardId}_${sequence}`
			}
		});
	}
	/**
	* Switch a streaming card to finalized state (streaming_mode: false).
	* Feishu auto-closes after 10min regardless, but callers should close
	* explicitly when producer completes.
	*
	* `summary` is optional; when provided, also updates the card's preview
	* text (the one shown in message lists / chat previews). Without it, the
	* preview stays at whatever was set during streaming (typically the
	* default "[Generating...]"), which looks stuck.
	*/
	async finishStreamingCard(cardId, sequence, summary) {
		const config = { streaming_mode: false };
		if (summary !== void 0) config.summary = { content: summary };
		await this.client.cardkit.v1.card.settings({
			path: { card_id: cardId },
			data: {
				settings: JSON.stringify({ config }),
				sequence,
				uuid: `s_${cardId}_${sequence}`
			}
		});
	}
	makeResult(ids) {
		return {
			messageId: ids[0],
			chunkIds: ids.length > 1 ? ids : void 0
		};
	}
};
/**
* Pick the reply target for chunk `i` of a split message.
*
* The first chunk uses the caller's `replyTo` verbatim. Continuation chunks
* chain off the previous chunk's id so a long reply stays in one thread / reply
* line instead of breaking out as top-level messages (issue #190) — but only
* when the send is "anchored" (a reply target or a thread). A plain fresh send
* keeps its continuation chunks as independent top-level messages.
*/
function replyTargetForChunk(i, opts, ids) {
	if (i === 0) return opts.replyTo;
	return opts.replyTo != null || opts.replyInThread === true ? ids[i - 1] : void 0;
}
function splitPlain(text, limit) {
	if (text.length <= limit) return [text];
	const out = [];
	for (let i = 0; i < text.length; i += limit) out.push(text.slice(i, i + limit));
	return out;
}
//#endregion
//#region src/safety/chat-pipeline.ts
/**
* Per-scope pipeline that does two things at once:
*   - Debounce-based batch aggregation (for IM messages)
*   - Strict serialization of all work (both batched flushes and direct
*     `run()` tasks) via a promise chain
*
* Supports two entry points:
*   - push(msg, handler):  enqueue for batched dispatch
*   - run(task):           run a one-shot async task serialized after any
*                          pending batch and previous tasks
*/
var ChatPipeline = class {
	config;
	serialOnly;
	buffer = [];
	bufferChars = 0;
	timer;
	tail = Promise.resolve();
	pendingHandler;
	/** True while a flush is enqueued/in-flight. Used by mergeWhileBusy. */
	busy = false;
	constructor(config, serialOnly) {
		this.config = config;
		this.serialOnly = serialOnly;
	}
	push(msg, handler) {
		this.buffer.push(msg);
		this.bufferChars += msg.content.length;
		this.pendingHandler ??= handler;
		if (this.config.mergeWhileBusy && this.busy) return;
		if (this.buffer.length >= this.config.maxMessages || this.bufferChars >= this.config.maxChars) {
			this.clearTimer();
			this.enqueueFlush();
			return;
		}
		if (this.config.delayMs <= 0 || this.serialOnly) {
			this.clearTimer();
			this.enqueueFlush();
			return;
		}
		this.clearTimer();
		const delay = this.bufferChars >= this.config.longThresholdChars ? this.config.longDelayMs : this.config.delayMs;
		this.timer = setTimeout(() => {
			this.timer = void 0;
			this.enqueueFlush();
		}, delay);
	}
	run(task) {
		if (this.buffer.length > 0) {
			this.clearTimer();
			this.enqueueFlush();
		}
		const next = this.tail.then(task, task);
		this.tail = next.then(() => void 0, () => void 0);
		return next;
	}
	async flushNow() {
		if (this.buffer.length > 0) {
			this.clearTimer();
			this.enqueueFlush();
		}
		await this.tail;
	}
	isIdle() {
		return this.buffer.length === 0 && !this.timer;
	}
	dispose() {
		this.clearTimer();
		this.buffer = [];
		this.pendingHandler = void 0;
	}
	clearTimer() {
		if (this.timer) {
			clearTimeout(this.timer);
			this.timer = void 0;
		}
	}
	enqueueFlush() {
		if (this.buffer.length === 0) return;
		const batch = this.buffer;
		const handler = this.pendingHandler;
		this.buffer = [];
		this.bufferChars = 0;
		this.pendingHandler = void 0;
		if (!handler) return;
		const dispatch = {
			message: mergeBatch(batch),
			sourceIds: batch.map((m) => m.messageId)
		};
		this.busy = true;
		const task = () => handler(dispatch);
		const next = this.tail.then(task, task);
		this.tail = next.then(() => this.onFlushSettled(), () => this.onFlushSettled());
	}
	/**
	* Called when a flush task settles. Clears the busy flag and, under
	* mergeWhileBusy, immediately re-flushes anything that accumulated while
	* the handler was running — as one merged batch.
	*/
	onFlushSettled() {
		this.busy = false;
		if (this.config.mergeWhileBusy && this.buffer.length > 0) {
			this.clearTimer();
			this.enqueueFlush();
		}
	}
};
var ChatPipelineManager = class {
	config;
	pipelines = /* @__PURE__ */ new Map();
	constructor(config) {
		this.config = config;
	}
	push(scope, msg, handler) {
		this.getOrCreate(scope, false).push(msg, handler);
	}
	run(scope, task) {
		return this.getOrCreate(scope, true).run(task);
	}
	getOrCreate(scope, serialOnly) {
		let p = this.pipelines.get(scope);
		if (!p) {
			p = new ChatPipeline(this.config, serialOnly);
			this.pipelines.set(scope, p);
		}
		return p;
	}
	async flushAll() {
		await Promise.all([...this.pipelines.values()].map((p) => p.flushNow()));
	}
	async dispose() {
		await this.flushAll();
		for (const p of this.pipelines.values()) p.dispose();
		this.pipelines.clear();
	}
};
/**
* Merge a batch of NormalizedMessages (all from the same chat) into a
* single representative message. Keeps the latest-arrival metadata and
* unions content / resources / mentions.
*/
function mergeBatch(batch) {
	if (batch.length === 1) return batch[0];
	const last = batch[batch.length - 1];
	const content = batch.map((m) => m.content).filter((c) => c && c.length > 0).join("\n\n");
	const resources = dedupBy(batch.flatMap((m) => m.resources), (r) => r.fileKey);
	const mentions = dedupBy(batch.flatMap((m) => m.mentions), (m) => m.openId ?? m.key);
	return {
		...last,
		content,
		resources,
		mentions,
		mentionAll: batch.some((m) => m.mentionAll),
		mentionedBot: batch.some((m) => m.mentionedBot)
	};
}
function dedupBy(items, key) {
	const seen = /* @__PURE__ */ new Set();
	const out = [];
	for (const item of items) {
		const k = key(item);
		if (k == null) {
			out.push(item);
			continue;
		}
		if (seen.has(k)) continue;
		seen.add(k);
		out.push(item);
	}
	return out;
}
//#endregion
//#region src/safety/types.ts
const DEFAULT_BATCH = {
	delayMs: 600,
	longThresholdChars: 1e3,
	longDelayMs: 2e3,
	maxMessages: 8,
	maxChars: 4e3,
	mergeWhileBusy: false
};
const DEFAULT_DEDUP = {
	ttl: 12 * 36e5,
	maxEntries: 5e3,
	sweepIntervalMs: 5 * 6e4,
	namespace: "channel:seen"
};
const DEFAULT_STALE_MS = 30 * 6e4;
const DEFAULT_LOCK_TTL_MS = 5 * 6e4;
function resolveBatchConfig(cfg) {
	const t = cfg?.batch?.text ?? {};
	return {
		delayMs: t.delayMs ?? DEFAULT_BATCH.delayMs,
		longThresholdChars: t.longThresholdChars ?? DEFAULT_BATCH.longThresholdChars,
		longDelayMs: t.longDelayMs ?? DEFAULT_BATCH.longDelayMs,
		maxMessages: t.maxMessages ?? DEFAULT_BATCH.maxMessages,
		maxChars: t.maxChars ?? DEFAULT_BATCH.maxChars,
		mergeWhileBusy: cfg?.chatQueue?.mergeWhileBusy ?? DEFAULT_BATCH.mergeWhileBusy
	};
}
//#endregion
//#region src/safety/dedup-cache.ts
/**
* Two-tier dedup cache: hot in-memory LRU + injectable long-term Cache.
* Memory tier serves 99% of queries sub-millisecond; long-term tier
* survives process restarts when a persistent Cache implementation is
* provided. Default node-sdk `internalCache` is memory-only — callers
* wanting cross-restart persistence inject their own Cache (e.g., Redis).
*/
var SeenCache = class {
	cache;
	memory = /* @__PURE__ */ new Map();
	sweeper;
	ttlMs;
	maxMem;
	ns;
	constructor(cache, opts = {}) {
		this.cache = cache;
		this.ttlMs = opts.ttlMs ?? DEFAULT_DEDUP.ttl;
		this.maxMem = opts.maxMemEntries ?? DEFAULT_DEDUP.maxEntries;
		this.ns = opts.namespace ?? DEFAULT_DEDUP.namespace;
		const sweepMs = opts.sweepMs ?? DEFAULT_DEDUP.sweepIntervalMs;
		this.sweeper = setInterval(() => this.sweep(), sweepMs);
		this.sweeper.unref?.();
	}
	async has(id) {
		const now = Date.now();
		const exp = this.memory.get(id);
		if (exp && exp > now) {
			this.memory.delete(id);
			this.memory.set(id, exp);
			return true;
		}
		if (await this.cache.get(id, { namespace: this.ns })) {
			this.memory.set(id, now + this.ttlMs);
			this.evictIfNeeded();
			return true;
		}
		return false;
	}
	async add(id) {
		const expireAt = Date.now() + this.ttlMs;
		this.memory.set(id, expireAt);
		this.evictIfNeeded();
		try {
			await this.cache.set(id, "1", expireAt, { namespace: this.ns });
		} catch {}
	}
	evictIfNeeded() {
		while (this.memory.size > this.maxMem) {
			const first = this.memory.keys().next().value;
			if (first === void 0) break;
			this.memory.delete(first);
		}
	}
	sweep() {
		const now = Date.now();
		for (const [k, v] of this.memory) if (v <= now) this.memory.delete(k);
	}
	dispose() {
		clearInterval(this.sweeper);
		this.memory.clear();
	}
};
//#endregion
//#region src/safety/policy-gate.ts
var PolicyGate = class {
	cfg;
	bot;
	constructor(cfg, bot) {
		this.cfg = { ...cfg ?? {} };
		this.bot = bot;
	}
	evaluate(msg) {
		if (msg.chatType === "group") return this.evaluateGroup(msg);
		return this.evaluateDm(msg);
	}
	evaluateGroup(msg) {
		const allow = this.cfg.groupAllowlist;
		if (allow && allow.length > 0 && !allow.includes(msg.chatId)) return {
			allowed: false,
			reason: "group_not_allowed"
		};
		if ((this.cfg.requireMention ?? true) && !msg.mentionedBot) return {
			allowed: false,
			reason: "no_mention"
		};
		if (msg.mentionAll && !(this.cfg.respondToMentionAll ?? false)) return {
			allowed: false,
			reason: "mention_all_blocked"
		};
		return { allowed: true };
	}
	evaluateDm(msg) {
		const mode = this.cfg.dmMode ?? "open";
		if (mode === "disabled") return {
			allowed: false,
			reason: "dm_disabled"
		};
		if (mode === "allowlist") {
			if (!(this.cfg.dmAllowlist ?? []).includes(msg.senderId)) return {
				allowed: false,
				reason: "sender_not_allowed"
			};
		}
		return { allowed: true };
	}
	updateConfig(partial) {
		this.cfg = {
			...this.cfg,
			...partial
		};
	}
	getConfig() {
		return this.cfg;
	}
	setBotIdentity(bot) {
		this.bot = bot;
	}
	getBotIdentity() {
		return this.bot;
	}
};
//#endregion
//#region src/safety/processing-lock.ts
/**
* Short-TTL in-memory lock to prevent concurrent processing of the same
* event — complements SeenCache by covering the "currently in flight"
* window, during which the event is not yet committed to SeenCache.
*/
var ProcessingLock = class {
	ttlMs;
	locks = /* @__PURE__ */ new Map();
	sweeper;
	constructor(ttlMs = DEFAULT_LOCK_TTL_MS, sweepMs = 6e4) {
		this.ttlMs = ttlMs;
		this.sweeper = setInterval(() => this.sweep(), sweepMs);
		this.sweeper.unref?.();
	}
	/** Returns true if the lock is acquired; false if already held. */
	acquire(id) {
		const now = Date.now();
		const exp = this.locks.get(id);
		if (exp && exp > now) return false;
		this.locks.set(id, now + this.ttlMs);
		return true;
	}
	release(id) {
		this.locks.delete(id);
	}
	sweep() {
		const now = Date.now();
		for (const [k, v] of this.locks) if (v <= now) this.locks.delete(k);
	}
	dispose() {
		clearInterval(this.sweeper);
		this.locks.clear();
	}
};
//#endregion
//#region src/safety/stale-detector.ts
function isStale(createTimeMs, windowMs = DEFAULT_STALE_MS) {
	if (!createTimeMs || !Number.isFinite(createTimeMs)) return false;
	return Date.now() - createTimeMs > windowMs;
}
//#endregion
//#region src/safety/index.ts
/**
* Pipeline entry facade for the channel's safety layer.
*
* Three tiers of protection, each targeting different event shapes:
*   - pushMessage: full pipeline (stale + dedup + policy + lock + batch + queue)
*   - pushAction:  dedup + lock + queue — for card button clicks and doc comments
*   - pushLight:   dedup only — for reactions
*/
var SafetyPipeline = class {
	seenCache;
	lock;
	policy;
	manager;
	staleWindow;
	queueEnabled;
	logger;
	onReject;
	onMessage;
	constructor(opts) {
		this.logger = opts.logger;
		this.onReject = opts.onReject;
		this.onMessage = opts.onMessage;
		this.staleWindow = opts.config?.staleMessageWindowMs ?? 18e5;
		this.queueEnabled = opts.config?.chatQueue?.enabled ?? true;
		this.seenCache = new SeenCache(opts.cache, {
			ttlMs: opts.config?.dedup?.ttl,
			maxMemEntries: opts.config?.dedup?.maxEntries,
			sweepMs: opts.config?.dedup?.sweepIntervalMs
		});
		this.lock = new ProcessingLock();
		this.policy = new PolicyGate(opts.policy, opts.botIdentity);
		this.manager = new ChatPipelineManager(resolveBatchConfig(opts.config));
	}
	async pushMessage(msg) {
		if (isStale(msg.createTime, this.staleWindow)) {
			this.logger.debug?.(`safety: drop stale message ${msg.messageId}`);
			return;
		}
		if (await this.seenCache.has(msg.messageId)) {
			this.logger.debug?.(`safety: drop duplicate message ${msg.messageId}`);
			return;
		}
		const decision = this.policy.evaluate(msg);
		if (!decision.allowed) {
			this.onReject({
				messageId: msg.messageId,
				chatId: msg.chatId,
				senderId: msg.senderId,
				reason: decision.reason ?? "group_not_allowed"
			});
			return;
		}
		if (!this.lock.acquire(msg.messageId)) {
			this.logger.debug?.(`safety: drop in-flight message ${msg.messageId}`);
			return;
		}
		const dispatchHandler = async (batch) => {
			try {
				await this.onMessage(batch.message);
			} catch (e) {
				this.logger.error?.(`safety: message handler threw`, e);
			} finally {
				for (const id of batch.sourceIds) {
					try {
						await this.seenCache.add(id);
					} catch {}
					this.lock.release(id);
				}
			}
		};
		if (this.queueEnabled) this.manager.push(msg.chatId, msg, dispatchHandler);
		else dispatchHandler({
			message: msg,
			sourceIds: [msg.messageId]
		});
	}
	async pushAction(eventId, queueScope, handler) {
		if (await this.seenCache.has(eventId)) {
			this.logger.debug?.(`safety: drop duplicate action ${eventId}`);
			return;
		}
		if (!this.lock.acquire(eventId)) {
			this.logger.debug?.(`safety: drop in-flight action ${eventId}`);
			return;
		}
		const task = async () => {
			try {
				return await handler();
			} catch (e) {
				this.logger.error?.(`safety: action handler threw`, e);
				return;
			} finally {
				try {
					await this.seenCache.add(eventId);
				} catch {}
				this.lock.release(eventId);
			}
		};
		if (this.queueEnabled) return this.manager.run(queueScope, task);
		return task();
	}
	async pushLight(eventId, handler) {
		if (await this.seenCache.has(eventId)) return;
		await this.seenCache.add(eventId);
		try {
			await handler();
		} catch (e) {
			this.logger.warn?.(`safety: light handler threw`, e);
		}
	}
	updatePolicy(partial) {
		this.policy.updateConfig(partial);
	}
	getPolicy() {
		return this.policy.getConfig();
	}
	setBotIdentity(bot) {
		this.policy.setBotIdentity(bot);
	}
	async dispose() {
		await this.manager.dispose();
		this.seenCache.dispose();
		this.lock.dispose();
	}
};
//#endregion
//#region src/channel.ts
var LarkChannel = class {
	rawClient;
	rawWsClient;
	botIdentity;
	/** Cloud-doc comment surface: fetch / reply / reactions with quirk fallbacks. */
	comments;
	opts;
	logger;
	dispatcher;
	handlers = {};
	connectPromise;
	connected = false;
	sender;
	safety;
	chatModeCache = new ChatModeCache();
	keepaliveHandle;
	proxyAgent;
	constructor(opts) {
		this.opts = opts;
		this.logger = new _larksuiteoapi_node_sdk.LoggerProxy(opts.loggerLevel ?? _larksuiteoapi_node_sdk.LoggerLevel.info, opts.logger ?? _larksuiteoapi_node_sdk.defaultLogger);
		this.rawClient = new _larksuiteoapi_node_sdk.Client({
			appId: opts.appId,
			appSecret: opts.appSecret,
			domain: opts.domain ?? _larksuiteoapi_node_sdk.Domain.Feishu,
			cache: opts.cache,
			httpInstance: opts.httpInstance,
			logger: opts.logger,
			loggerLevel: opts.loggerLevel,
			source: opts.source,
			extraUaTags: ["channel"]
		});
		this.dispatcher = new _larksuiteoapi_node_sdk.EventDispatcher({
			verificationToken: opts.webhook?.verificationToken,
			encryptKey: opts.webhook?.encryptKey,
			cache: opts.cache,
			logger: opts.logger,
			loggerLevel: opts.loggerLevel
		});
		this.sender = new OutboundSender(this.rawClient, opts.outbound ?? {}, this.logger);
		this.comments = new CommentSurface(this.rawClient, this.logger);
		this.configureHttp();
		this.safety = new SafetyPipeline({
			config: opts.safety,
			policy: opts.policy,
			cache: opts.cache ?? _larksuiteoapi_node_sdk.internalCache,
			logger: this.logger,
			onReject: (evt) => {
				this.handlers.reject?.(evt);
			},
			onMessage: async (merged) => {
				const handler = this.handlers.message;
				if (handler) await handler(merged);
			}
		});
	}
	async connect() {
		if (this.connectPromise) return this.connectPromise;
		this.connectPromise = this.doConnect().catch((err) => {
			this.connectPromise = void 0;
			throw err;
		});
		return this.connectPromise;
	}
	async doConnect() {
		this.botIdentity = await this.fetchBotIdentity();
		this.safety.setBotIdentity(this.botIdentity);
		this.registerDispatcherHandlers();
		if ((this.opts.transport ?? "websocket") === "websocket") {
			await this.connectWebSocket(15e3);
			this.startKeepaliveIfEnabled();
		}
		this.connected = true;
	}
	startKeepaliveIfEnabled() {
		if (!this.opts.keepalive?.enabled || this.keepaliveHandle) return;
		this.keepaliveHandle = startKeepalive({
			getConnectionStatus: () => this.getConnectionStatus(),
			domain: String(this.opts.domain ?? _larksuiteoapi_node_sdk.Domain.Feishu),
			forceReconnect: () => this.forceReconnect(),
			onUnrecoverable: this.opts.keepalive.onUnrecoverable,
			logger: this.logger,
			intervalMs: this.opts.keepalive.intervalMs
		});
	}
	/**
	* Tear down the current WebSocket and re-establish it. Used by the
	* keepalive watchdog when the connection looks stuck. Throws if the fresh
	* handshake fails, so keepalive can surface it via `onUnrecoverable`.
	*/
	async forceReconnect() {
		try {
			this.rawWsClient?.close({ force: true });
		} catch {}
		this.rawWsClient = void 0;
		await this.connectWebSocket(this.opts.handshakeTimeoutMs ?? 15e3);
	}
	/**
	* Apply a per-request timeout and/or proxy to node-sdk's shared
	* `defaultHttpInstance` (a typed `AxiosInstance` — `defaults` is visible,
	* no cast needed). Only runs when the caller opted in, and only when no
	* custom `httpInstance` was supplied: a caller who brings their own HTTP
	* instance owns its configuration, and we don't mutate a process-wide
	* singleton behind their back.
	*/
	configureHttp() {
		if (this.opts.httpInstance) return;
		const timeout = this.opts.httpTimeoutMs;
		const proxyAgent = this.opts.respectProxyEnv ? this.proxyAgentFromEnv() : void 0;
		if (timeout == null && !proxyAgent) return;
		if (timeout != null) _larksuiteoapi_node_sdk.defaultHttpInstance.defaults.timeout = timeout;
		if (proxyAgent) {
			_larksuiteoapi_node_sdk.defaultHttpInstance.defaults.httpsAgent = proxyAgent;
			_larksuiteoapi_node_sdk.defaultHttpInstance.defaults.httpAgent = proxyAgent;
		}
	}
	/**
	* Resolve the Node http(s) agent for the WebSocket transport: an explicit
	* `agent` option wins; otherwise, when `respectProxyEnv` is set, build one
	* from `HTTPS_PROXY` / `HTTP_PROXY`.
	*/
	resolveWsAgent() {
		if (this.opts.agent) return this.opts.agent;
		if (!this.opts.respectProxyEnv) return void 0;
		return this.proxyAgentFromEnv();
	}
	/** Lazily build (and cache) a proxy agent from the proxy env vars. Shared
	*  by the WebSocket transport and the REST HTTP instance. */
	proxyAgentFromEnv() {
		if (this.proxyAgent) return this.proxyAgent;
		const proxyUrl = process.env.HTTPS_PROXY ?? process.env.https_proxy ?? process.env.HTTP_PROXY ?? process.env.http_proxy;
		if (!proxyUrl) return void 0;
		this.proxyAgent = new https_proxy_agent.HttpsProxyAgent(proxyUrl);
		this.logger.info?.("channel: proxy detected", { proxy: redactProxyUrl(proxyUrl) });
		return this.proxyAgent;
	}
	/**
	* Construct the underlying WSClient and wait for its `onReady` callback —
	* so `connect()` only resolves after the first WebSocket handshake
	* actually succeeds. Rejects on `onError` or if the handshake doesn't
	* complete within `timeoutMs`.
	*
	* Also wires `onReconnecting` / `onReconnected` callbacks to emit the
	* corresponding public events.
	*/
	connectWebSocket(timeoutMs) {
		return new Promise((resolve, reject) => {
			let settled = false;
			const timer = setTimeout(() => {
				if (settled) return;
				settled = true;
				reject(new LarkChannelError("not_connected", `WebSocket handshake did not complete within ${timeoutMs}ms`));
			}, timeoutMs);
			this.rawWsClient = new _larksuiteoapi_node_sdk.WSClient({
				appId: this.opts.appId,
				appSecret: this.opts.appSecret,
				domain: this.opts.domain ?? _larksuiteoapi_node_sdk.Domain.Feishu,
				logger: this.opts.logger,
				loggerLevel: this.opts.loggerLevel,
				httpInstance: this.opts.httpInstance,
				autoReconnect: true,
				source: this.opts.source,
				extraUaTags: ["channel"],
				agent: this.resolveWsAgent(),
				wsConfig: this.opts.wsConfig,
				handshakeTimeoutMs: this.opts.handshakeTimeoutMs,
				onReady: () => {
					if (settled) return;
					settled = true;
					clearTimeout(timer);
					resolve();
				},
				onError: (err) => {
					if (settled) return;
					settled = true;
					clearTimeout(timer);
					reject(new LarkChannelError("not_connected", `WebSocket connect failed: ${err.message}`, { cause: err }));
				},
				onReconnecting: () => this.handlers.reconnecting?.(),
				onReconnected: () => this.handlers.reconnected?.()
			});
			this.rawWsClient.start({ eventDispatcher: this.dispatcher });
		});
	}
	async disconnect() {
		if (!this.connected) return;
		this.keepaliveHandle?.stop();
		this.keepaliveHandle = void 0;
		try {
			this.rawWsClient?.close({});
		} catch {}
		try {
			await this.safety.dispose();
		} catch {}
		this.connected = false;
		this.connectPromise = void 0;
	}
	/**
	* Snapshot of the WebSocket lifecycle (state, last/next connect times,
	* current reconnect attempts). Returns `undefined` when the channel
	* hasn't initialized a WSClient yet (e.g., before `connect()` is called
	* or under the webhook transport).
	*/
	getConnectionStatus() {
		return this.rawWsClient?.getConnectionStatus();
	}
	on(nameOrMap, handler) {
		if (typeof nameOrMap === "string") return this.attachSingle(nameOrMap, handler);
		const unsubs = [];
		Object.keys(nameOrMap).forEach((k) => {
			const fn = nameOrMap[k];
			if (fn) unsubs.push(this.attachSingle(k, fn));
		});
		return () => {
			unsubs.forEach((u) => {
				u();
			});
		};
	}
	attachSingle(name, handler) {
		if (this.handlers[name]) this.logger.warn(`channel: handler for "${name}" is being overwritten`);
		this.handlers[name] = handler;
		return () => {
			if (this.handlers[name] === handler) delete this.handlers[name];
		};
	}
	async send(to, input, opts) {
		return this.sender.send(to, input, opts);
	}
	async stream(to, input, opts) {
		return this.sender.stream(to, input, opts);
	}
	async updateCard(messageId, card) {
		await this.sender.patchCard(messageId, card);
	}
	/**
	* Create a standalone CardKit 2.0 card entity (`cardkit.v1.card.create`) and
	* return its `card_id`. The card isn't attached to any message yet — send a
	* message that references it via `channel.send(to, { cardId })`, then drive
	* it with {@link updateCardById}. This is the managed-card lifecycle: one
	* entity, many in-place updates, decoupled from the message that displays it.
	*/
	async createCard(cardJson) {
		return { cardId: await this.sender.createCardInstance(cardJson) };
	}
	/**
	* Full-content update of a card entity by `card_id` (`cardkit.v1.card.update`).
	* `sequence` must strictly increase across calls for the same card — Feishu
	* rejects stale/out-of-order sequences so a slow update can't overwrite a
	* newer one. Unlike {@link updateCard} (which targets a message_id), this
	* updates the shared entity, so every message referencing the card_id
	* re-renders.
	*/
	async updateCardById(cardId, cardJson, sequence) {
		await this.sender.updateCardFull(cardId, cardJson, sequence);
	}
	/**
	* Edit an already-sent message's text/post content. Uses `im.v1.message.update`
	* which (per Feishu docs) only supports editing text and rich-text (post)
	* messages. For cards, use {@link updateCard} instead — a wrong attempt to
	* use this on a card would hit the same API and fail with a clearer
	* Feishu-side error.
	*/
	async editMessage(messageId, text) {
		await this.rawClient.im.v1.message.update({
			path: { message_id: messageId },
			data: {
				msg_type: "text",
				content: JSON.stringify({ text })
			}
		});
	}
	async recallMessage(messageId) {
		await this.rawClient.im.v1.message.delete({ path: { message_id: messageId } });
	}
	/**
	* Add an emoji reaction to a message. Returns the `reaction_id` Feishu
	* assigned — stash it if you want to {@link removeReaction} later,
	* since the raw `im.message.reaction.*_v1` events don't carry the id.
	* Only the bot's own reactions can be removed.
	*/
	async addReaction(messageId, emojiType) {
		const r = await this.rawClient.im.v1.messageReaction.create({
			path: { message_id: messageId },
			data: { reaction_type: { emoji_type: emojiType } }
		});
		const rid = r?.data?.reaction_id ?? r?.reaction_id;
		if (!rid) throw new LarkChannelError("unknown", "messageReaction.create returned no reaction_id");
		return rid;
	}
	/**
	* Remove a reaction by its `reaction_id` (the value returned from
	* {@link addReaction}). Only the bot's own reactions can be removed —
	* removing a user-added reaction will fail with a Feishu permission
	* error.
	*/
	async removeReaction(messageId, reactionId) {
		await this.rawClient.im.v1.messageReaction.delete({ path: {
			message_id: messageId,
			reaction_id: reactionId
		} });
	}
	/**
	* Convenience: remove the bot's reaction on `messageId` matching
	* `emojiType`, without needing the `reaction_id`. Lists the message's
	* reactions filtered by emoji, picks the one added by this bot
	* (operator_type === 'app'), and deletes it. Returns `true` if a
	* matching reaction was found and deleted, `false` otherwise (including
	* the case where the bot never added that emoji).
	*/
	async removeReactionByEmoji(messageId, emojiType) {
		const mine = ((await this.rawClient.im.v1.messageReaction.list({
			path: { message_id: messageId },
			params: {
				reaction_type: emojiType,
				page_size: 50
			}
		}))?.data?.items ?? []).find((it) => it.operator?.operator_type === "app");
		if (!mine?.reaction_id) return false;
		await this.removeReaction(messageId, mine.reaction_id);
		return true;
	}
	/**
	* Download a resource (image / file / audio / video / sticker) carried by a
	* **received** message. Feishu serves message resources via
	* `im.v1.messageResource.get`, which needs both the owning `messageId` and
	* the resource's `fileKey` — the `im/v1/images|files/:key` endpoints only
	* work for media the app itself uploaded and return 400 for received media.
	*
	* `type` is `'image'` for image resources, `'file'` for everything else
	* (file / audio / video / sticker) — matching `ResourceDescriptor.type`.
	*/
	async downloadResource(messageId, fileKey, type) {
		const { buffer } = await this.downloadResourceWithMeta(messageId, fileKey, type);
		return buffer;
	}
	/**
	* Like {@link downloadResource}, but also returns the server's response
	* `content-type` (when present). Feishu's `im.v1.messageResource.get`
	* carries the resource's real MIME in the response headers — needed to pick
	* an accurate file extension. `contentType` is the media type with any
	* parameters (e.g. `; charset=...`) stripped, or `undefined` when the header
	* is absent (e.g. a defensive raw-`Buffer` response). Callers should fall
	* back to a per-kind default in that case.
	*/
	async downloadResourceWithMeta(messageId, fileKey, type) {
		const r = await this.rawClient.im.v1.messageResource.get({
			path: {
				message_id: messageId,
				file_key: fileKey
			},
			params: { type }
		});
		return {
			buffer: await bufferFromStream(r),
			contentType: extractContentType(r)
		};
	}
	/**
	* Stream a message resource straight to `destPath` without ever holding the
	* whole payload in memory — the HTTP response is `pipe`d to the file, so a
	* 100 MB attachment costs only stream-buffer overhead, not 100 MB of JS
	* heap. Prefer this over {@link downloadResource} /
	* {@link downloadResourceWithMeta} whenever the bytes are headed for disk
	* (e.g. a size-limited attachment cache): those materialize a full `Buffer`
	* via `Buffer.concat`, which can OOM the process when several large
	* downloads run concurrently.
	*
	* Returns the server `content-type` (params stripped; `undefined` when
	* absent) for MIME/extension detection, and the number of bytes written.
	* The parent directory of `destPath` must already exist.
	*/
	async downloadResourceToFile(messageId, fileKey, type, destPath) {
		const r = await this.rawClient.im.v1.messageResource.get({
			path: {
				message_id: messageId,
				file_key: fileKey
			},
			params: { type }
		});
		return {
			contentType: extractContentType(r),
			bytesWritten: await streamToFile(r, destPath)
		};
	}
	/**
	* Create a group chat (`im.v1.chat.create`) and return its `chat_id`.
	* `inviteUserIds` seeds the membership; the ids are interpreted per
	* `userIdType` (default `'open_id'`). Requires the `im:chat` scope.
	*/
	async createChat(opts) {
		const chatId = (await this.rawClient.im.v1.chat.create({
			params: { user_id_type: opts.userIdType ?? "open_id" },
			data: {
				name: opts.name,
				description: opts.description,
				chat_mode: opts.chatMode ?? "group",
				chat_type: opts.chatType ?? "private",
				user_id_list: opts.inviteUserIds
			}
		})).data?.chat_id;
		if (!chatId) throw new LarkChannelError("unknown", "im.v1.chat.create returned no chat_id");
		return { chatId };
	}
	/**
	* List the chats this bot is a member of (`im.v1.chat.list`), following
	* pagination automatically. `pageSize` is clamped to Feishu's max of 100;
	* `maxPages` caps how many pages are fetched (default 10) so an account in
	* thousands of chats can't spin forever. Returns `{ id, name }` per chat.
	*/
	async listChats(opts) {
		const pageSize = Math.min(Math.max(opts?.pageSize ?? 100, 1), 100);
		const maxPages = opts?.maxPages ?? 10;
		const out = [];
		let pageToken;
		for (let page = 0; page < maxPages; page++) {
			const d = (await this.rawClient.im.v1.chat.list({ params: {
				page_size: pageSize,
				page_token: pageToken
			} }))?.data;
			for (const it of d?.items ?? []) if (it.chat_id) out.push({
				id: it.chat_id,
				name: it.name ?? ""
			});
			if (!d?.has_more || !d.page_token) break;
			pageToken = d.page_token;
		}
		return out;
	}
	/**
	* Fetch this app's own metadata (`application.v6.application.get`) — the
	* `app_id` is the one the channel was constructed with, so callers don't
	* pass it. Primarily used to resolve the app owner/admin (`ownerId`) for
	* access control. Requires the application-info scope.
	*/
	async getAppInfo(opts) {
		const app = (await this.rawClient.application.v6.application.get({
			path: { app_id: this.opts.appId },
			params: {
				lang: opts?.lang ?? "zh_cn",
				user_id_type: opts?.userIdType ?? "open_id"
			}
		})).data?.app;
		return {
			ownerId: app?.owner?.owner_id,
			appName: app?.app_name
		};
	}
	async getChatInfo(chatId) {
		const d = (await this.rawClient.im.v1.chat.get({ path: { chat_id: chatId } })).data ?? {};
		return {
			chatId,
			name: d.name,
			description: d.description,
			chatType: d.chat_mode ?? "group",
			ownerId: d.owner_id,
			memberCount: d.user_count
		};
	}
	/**
	* Fetch the chat's mode via `im.v1.chat.get`. Returns one of:
	*   - 'p2p'   — direct (1:1) chat
	*   - 'group' — ordinary group
	*   - 'topic' — topic group
	*
	* Unknown / missing values fall back to 'group' for consistency with
	* {@link getChatInfo}. The underlying API call is not cached — chat
	* mode rarely changes within a chat's lifetime, so callers that read
	* this on every inbound message should keep their own cache keyed by
	* `chatId`.
	*
	* Throws on API failure (network, permission, invalid chatId) so the
	* caller can decide how to handle it; silently defaulting would hide
	* real problems.
	*/
	async getChatMode(chatId) {
		const mode = (await this.rawClient.im.v1.chat.get({ path: { chat_id: chatId } })).data?.chat_mode;
		if (mode === "p2p") return "p2p";
		if (mode === "topic") return "topic";
		return "group";
	}
	/**
	* Fetch a message by id and return it as a {@link NormalizedMessage} — the
	* same shape live `message` events produce. Useful for resolving a
	* reply-quoted message: `im.v1.message.get` returns a flat item list
	* (parent + descendants for merge_forward), which this method feeds back
	* through {@link normalize} so merge_forward gets the same
	* `<forwarded_messages>` expansion as live events.
	*
	* Sunk from bridge's `quote.ts`, which previously synthesized a fake raw
	* event and called the internal `normalize()` directly. Returns
	* `undefined` when the message can't be fetched or has no parent item.
	* `stripBotMentions` is off here so the raw quoted content is preserved.
	*/
	/**
	* Fetch a message's raw `data.items[]` (`im.v1.message.get`) without running
	* them through {@link normalize} — for callers that need fidelity the
	* normalizer drops: original `body.content` JSON, `mentions`, `sender.id`,
	* `create_time`. For merge_forward the list is the parent followed by its
	* descendants (each carrying `upper_message_id`).
	*
	* `cardContentType` maps to the `card_msg_content_type` query param.
	* Defaults to `'user_card_content'` so interactive messages return the
	* original CardKit 2.0 card JSON (`user_dsl`) rather than the v1-canonical
	* downgrade. Pass `null` to omit the param entirely.
	*/
	async fetchRawMessage(messageId, opts) {
		const cardContentType = opts?.cardContentType === void 0 ? "user_card_content" : opts.cardContentType;
		return (await this.rawClient.im.v1.message.get({
			path: { message_id: messageId },
			params: cardContentType ? { card_msg_content_type: cardContentType } : void 0
		}))?.data?.items ?? [];
	}
	async fetchMessage(messageId) {
		let items;
		try {
			items = (await this.rawClient.im.v1.message.get({ path: { message_id: messageId } }))?.data?.items ?? [];
		} catch (e) {
			this.logger.warn?.("channel: fetchMessage failed", e);
			return;
		}
		const parent = items[0];
		if (!parent || !parent.message_id) return void 0;
		const fetchSubMessages = async (mid) => {
			if (mid === parent.message_id) return items;
			try {
				return (await this.rawClient.im.v1.message.get({ path: { message_id: mid } }))?.data?.items ?? [];
			} catch {
				return [];
			}
		};
		const fakeRaw = {
			sender: { sender_id: { open_id: parent.sender?.id } },
			message: {
				message_id: parent.message_id,
				chat_id: "",
				chat_type: "group",
				message_type: parent.msg_type ?? "text",
				content: parent.body?.content ?? "",
				create_time: parent.create_time !== void 0 ? String(parent.create_time) : void 0,
				mentions: parent.mentions
			}
		};
		try {
			return await normalize(fakeRaw, {
				botIdentity: this.botIdentity ?? {
					openId: "",
					name: ""
				},
				fetchSubMessages,
				stripBotMentions: false
			});
		} catch (e) {
			this.logger.warn?.("channel: fetchMessage normalize failed", e);
			return;
		}
	}
	updatePolicy(partial) {
		this.safety.updatePolicy(partial);
	}
	getPolicy() {
		return this.safety.getPolicy();
	}
	async fetchBotIdentity() {
		let lastError;
		try {
			const r = await this.rawClient.request({
				url: "/open-apis/bot/v3/info",
				method: "GET"
			});
			const bot = r.bot;
			if (bot?.open_id) return {
				openId: bot.open_id,
				name: bot.app_name ?? "bot"
			};
			lastError = /* @__PURE__ */ new Error(`bot/v3/info response missing open_id: ${JSON.stringify(r).slice(0, 200)}`);
		} catch (e) {
			lastError = e;
		}
		const classified = classifyError(lastError);
		throw new LarkChannelError(classified.code === "unknown" ? "not_connected" : classified.code, "could not resolve bot identity via /open-apis/bot/v3/info — required for channel to function", { cause: lastError });
	}
	registerDispatcherHandlers() {
		const fetchSubMessages = async (mid) => {
			try {
				return (await this.rawClient.im.v1.message.get({ path: { message_id: mid } })).data?.items ?? [];
			} catch (e) {
				this.logger.warn?.("channel: fetchSubMessages failed", e);
				return [];
			}
		};
		const includeRaw = this.opts.includeRawEvent ?? this.opts.includeRawInMessage ?? false;
		const normalizeOpts = {
			botIdentity: this.botIdentity,
			stripBotMentions: true,
			includeRaw,
			fetchSubMessages
		};
		this.dispatcher.register({
			"im.message.receive_v1": async (raw) => {
				try {
					const msg = await normalize(raw, normalizeOpts);
					if (this.opts.resolveChatMode) msg.chatMode = await this.chatModeCache.resolve(msg.chatId, (id) => this.getChatMode(id));
					await this.safety.pushMessage(msg);
				} catch (e) {
					this.emitError(e);
				}
			},
			"card.action.trigger": async (raw) => {
				const evt = normalizeCardAction(raw, { includeRaw });
				if (!evt) return void 0;
				const actionId = cardActionId(evt.action);
				return this.safety.pushAction(`card:${evt.messageId}:${evt.operator.openId}:${actionId}`, evt.chatId, async () => {
					const h = this.handlers.cardAction;
					return h ? h(evt) : void 0;
				});
			},
			"im.message.reaction.created_v1": async (raw) => {
				const evt = normalizeReaction(raw, "added", { includeRaw });
				if (!evt) return;
				const key = reactionKey(evt);
				await this.safety.pushLight(key, () => this.handlers.reaction?.(evt));
			},
			"im.message.reaction.deleted_v1": async (raw) => {
				const evt = normalizeReaction(raw, "removed", { includeRaw });
				if (!evt) return;
				const key = reactionKey(evt);
				await this.safety.pushLight(key, () => this.handlers.reaction?.(evt));
			},
			"im.chat.member.bot.added_v1": (raw) => {
				const evt = normalizeBotAdded(raw, { includeRaw });
				if (!evt) return;
				try {
					this.handlers.botAdded?.(evt);
				} catch (e) {
					this.emitError(e);
				}
			},
			"drive.notice.comment_add_v1": async (raw) => {
				const evt = normalizeComment(raw, { includeRaw });
				if (!evt) return;
				await this.safety.pushAction(`comment:${evt.fileToken}:${evt.commentId}:${evt.replyId ?? ""}`, evt.fileToken, async () => {
					const h = this.handlers.comment;
					if (h) await h(evt);
				});
			}
		});
	}
	emitError(e) {
		const err = e instanceof LarkChannelError ? e : new LarkChannelError("unknown", String(e?.message ?? e), { cause: e });
		const handler = this.handlers.error;
		if (handler) handler(err);
		else this.logger.error?.("channel: unhandled error", err);
	}
};
function createLarkChannel(opts) {
	return new LarkChannel(opts);
}
/** Mask `user:pass@` credentials in a proxy URL before logging it. */
function redactProxyUrl(url) {
	return url.replace(/\/\/[^:@/]+:[^@/]+@/, "//[redacted]@");
}
/**
* Pull the `content-type` media type out of a download response's headers.
* The code-gen download endpoints expose axios response headers on the
* wrapper object; header names are case-insensitive, so check both casings.
* Strips any `; charset=…` / `; boundary=…` parameters and returns the bare
* media type (lowercased), or `undefined` when no usable header is present.
*/
function extractContentType(raw) {
	if (typeof raw !== "object" || raw === null) return void 0;
	const headers = raw.headers;
	if (!headers) return void 0;
	const value = headers["content-type"] ?? headers["Content-Type"];
	if (typeof value !== "string") return void 0;
	return value.split(";", 1)[0]?.trim().toLowerCase() || void 0;
}
/**
* Stream a download response to disk without buffering it in heap. The
* code-gen download endpoints expose the body via `getReadableStream()`,
* which we `pipeline` straight into a write stream (back-pressure aware,
* cleans up on error). Falls back to writing an already-materialized
* `Buffer` / `Uint8Array` for the defensive non-stream response shapes —
* those are already in memory, so there's nothing to stream. Returns the
* number of bytes written.
*/
async function streamToFile(raw, destPath) {
	if (typeof raw === "object" && raw !== null) {
		const r = raw;
		if (typeof r.getReadableStream === "function") {
			await (0, node_stream_promises.pipeline)(r.getReadableStream(), (0, node_fs.createWriteStream)(destPath));
			const { size } = await (0, node_fs_promises.stat)(destPath);
			return size;
		}
		if (Buffer.isBuffer(r.data)) {
			await (0, node_fs_promises.writeFile)(destPath, r.data);
			return r.data.length;
		}
		if (r.data instanceof Uint8Array) {
			const buf = Buffer.from(r.data);
			await (0, node_fs_promises.writeFile)(destPath, buf);
			return buf.length;
		}
	}
	if (Buffer.isBuffer(raw)) {
		await (0, node_fs_promises.writeFile)(destPath, raw);
		return raw.length;
	}
	if (raw instanceof Uint8Array) {
		const buf = Buffer.from(raw);
		await (0, node_fs_promises.writeFile)(destPath, buf);
		return buf.length;
	}
	throw new LarkChannelError("unknown", "unexpected download response type");
}
async function bufferFromStream(raw) {
	if (Buffer.isBuffer(raw)) return raw;
	if (raw instanceof Uint8Array) return Buffer.from(raw);
	if (typeof raw === "object" && raw !== null) {
		const r = raw;
		if (typeof r.getReadableStream === "function") return await readableToBuffer(r.getReadableStream());
		if (Buffer.isBuffer(r.data)) return r.data;
		if (r.data instanceof Uint8Array) return Buffer.from(r.data);
	}
	throw new LarkChannelError("unknown", "unexpected download response type");
}
function readableToBuffer(stream) {
	return new Promise((resolve, reject) => {
		const chunks = [];
		stream.on("data", (chunk) => {
			chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk));
		});
		stream.on("end", () => resolve(Buffer.concat(chunks)));
		stream.on("error", reject);
	});
}
function reactionKey(evt) {
	return `rx:${evt.messageId}:${evt.operator.openId}:${evt.emojiType}:${evt.action}:${evt.actionTime ?? 0}`;
}
/**
* Build a stable identity for a card action event's button/element, so that
* different clicks on the same card by the same user dedup independently.
* `tag` plus serialized `value` is enough to tell buttons apart; `name` and
* `option` are rolled in for form-style interactions where the same value
* may repeat but the triggering element differs. The serialized payload is
* length-clamped to keep cache keys small.
*/
function cardActionId(action) {
	const serialized = typeof action.value === "string" ? action.value : JSON.stringify(action.value ?? "");
	const valuePart = serialized.length > 128 ? serialized.slice(0, 128) : serialized;
	return `${action.tag}|${action.name ?? ""}|${action.option ?? ""}|${valuePart}`;
}
//#endregion
exports.CommentSurface = CommentSurface;
exports.LarkChannel = LarkChannel;
exports.LarkChannelError = LarkChannelError;
exports.createLarkChannel = createLarkChannel;
exports.normalize = normalize;
exports.normalizeBotAdded = normalizeBotAdded;
exports.normalizeCardAction = normalizeCardAction;
exports.normalizeComment = normalizeComment;
exports.normalizeReaction = normalizeReaction;
exports.registerApp = _larksuiteoapi_node_sdk.registerApp;
